<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>BrainAlign.came.utils package &mdash; BrainAlign 0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            BrainAlign
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">BrainAlign.came.utils package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-BrainAlign.came.utils.analyze">BrainAlign.came.utils.analyze module</a><ul>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.abstract_nodes"><code class="docutils literal notranslate"><span class="pre">abstract_nodes()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.abstract_ov_edges"><code class="docutils literal notranslate"><span class="pre">abstract_ov_edges()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.abstract_vv_edges"><code class="docutils literal notranslate"><span class="pre">abstract_vv_edges()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.adata_subgroup_edges"><code class="docutils literal notranslate"><span class="pre">adata_subgroup_edges()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.adj_to_edges"><code class="docutils literal notranslate"><span class="pre">adj_to_edges()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.aggregate_links"><code class="docutils literal notranslate"><span class="pre">aggregate_links()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.arrange_contingency_mat"><code class="docutils literal notranslate"><span class="pre">arrange_contingency_mat()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.compare_deg_dicts"><code class="docutils literal notranslate"><span class="pre">compare_deg_dicts()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.compare_degs_adata"><code class="docutils literal notranslate"><span class="pre">compare_degs_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.compare_degs_seurat"><code class="docutils literal notranslate"><span class="pre">compare_degs_seurat()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.compare_modules"><code class="docutils literal notranslate"><span class="pre">compare_modules()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.compute_common_private"><code class="docutils literal notranslate"><span class="pre">compute_common_private()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.edges_from_adata"><code class="docutils literal notranslate"><span class="pre">edges_from_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.export_neighbor_subgraph_df"><code class="docutils literal notranslate"><span class="pre">export_neighbor_subgraph_df()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.export_subgraph_df"><code class="docutils literal notranslate"><span class="pre">export_subgraph_df()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.get_adata_neighbors"><code class="docutils literal notranslate"><span class="pre">get_adata_neighbors()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.load_dpair_and_model"><code class="docutils literal notranslate"><span class="pre">load_dpair_and_model()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.make_abstracted_graph"><code class="docutils literal notranslate"><span class="pre">make_abstracted_graph()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.make_nx_input_from_df"><code class="docutils literal notranslate"><span class="pre">make_nx_input_from_df()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.module_enrichment_for_classes"><code class="docutils literal notranslate"><span class="pre">module_enrichment_for_classes()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.module_homo_weights"><code class="docutils literal notranslate"><span class="pre">module_homo_weights()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.neighbor_induced_edgedf"><code class="docutils literal notranslate"><span class="pre">neighbor_induced_edgedf()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.neighbor_induced_subgraph"><code class="docutils literal notranslate"><span class="pre">neighbor_induced_subgraph()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.nx_from_adata"><code class="docutils literal notranslate"><span class="pre">nx_from_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.nx_multipartite_graph"><code class="docutils literal notranslate"><span class="pre">nx_multipartite_graph()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.nx_neighbors"><code class="docutils literal notranslate"><span class="pre">nx_neighbors()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.nx_to_dfs"><code class="docutils literal notranslate"><span class="pre">nx_to_dfs()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.nx_to_nodedf"><code class="docutils literal notranslate"><span class="pre">nx_to_nodedf()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.set_precomputed_neighbors"><code class="docutils literal notranslate"><span class="pre">set_precomputed_neighbors()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.weight_linked_vars"><code class="docutils literal notranslate"><span class="pre">weight_linked_vars()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.weight_linked_vars_by_expr"><code class="docutils literal notranslate"><span class="pre">weight_linked_vars_by_expr()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.weight_normalize_by_size"><code class="docutils literal notranslate"><span class="pre">weight_normalize_by_size()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.wrapper_confus_mat"><code class="docutils literal notranslate"><span class="pre">wrapper_confus_mat()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.analyze.wrapper_contingency_mat"><code class="docutils literal notranslate"><span class="pre">wrapper_contingency_mat()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-BrainAlign.came.utils.base">BrainAlign.came.utils.base module</a><ul>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.check_dirs"><code class="docutils literal notranslate"><span class="pre">check_dirs()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.dec_timewrapper"><code class="docutils literal notranslate"><span class="pre">dec_timewrapper()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.dict_struct"><code class="docutils literal notranslate"><span class="pre">dict_struct()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.dict_to_pairs"><code class="docutils literal notranslate"><span class="pre">dict_to_pairs()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.load_json_dict"><code class="docutils literal notranslate"><span class="pre">load_json_dict()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.load_pickle"><code class="docutils literal notranslate"><span class="pre">load_pickle()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.make_nowtime_tag"><code class="docutils literal notranslate"><span class="pre">make_nowtime_tag()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.make_pairs_from_lists"><code class="docutils literal notranslate"><span class="pre">make_pairs_from_lists()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.map_by_sme"><code class="docutils literal notranslate"><span class="pre">map_by_sme()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.pairs_to_dict"><code class="docutils literal notranslate"><span class="pre">pairs_to_dict()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.save_json_dict"><code class="docutils literal notranslate"><span class="pre">save_json_dict()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.save_pickle"><code class="docutils literal notranslate"><span class="pre">save_pickle()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.split_df"><code class="docutils literal notranslate"><span class="pre">split_df()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.strline"><code class="docutils literal notranslate"><span class="pre">strline()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.subsample_each_group"><code class="docutils literal notranslate"><span class="pre">subsample_each_group()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.subsample_single"><code class="docutils literal notranslate"><span class="pre">subsample_single()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.base.write_info"><code class="docutils literal notranslate"><span class="pre">write_info()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-BrainAlign.came.utils.downsample_counts">BrainAlign.came.utils.downsample_counts module</a><ul>
<li><a class="reference internal" href="#BrainAlign.came.utils.downsample_counts.downsample_counts"><code class="docutils literal notranslate"><span class="pre">downsample_counts()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.downsample_counts.downsample_counts_per_cell"><code class="docutils literal notranslate"><span class="pre">downsample_counts_per_cell()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.downsample_counts.downsample_total_counts"><code class="docutils literal notranslate"><span class="pre">downsample_total_counts()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-BrainAlign.came.utils.evaluation">BrainAlign.came.utils.evaluation module</a><ul>
<li><a class="reference internal" href="#BrainAlign.came.utils.evaluation.accuracy"><code class="docutils literal notranslate"><span class="pre">accuracy()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.evaluation.get_AMI"><code class="docutils literal notranslate"><span class="pre">get_AMI()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.evaluation.get_F1_score"><code class="docutils literal notranslate"><span class="pre">get_F1_score()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-BrainAlign.came.utils.plot">BrainAlign.came.utils.plot module</a><ul>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.adata_embed_with_values"><code class="docutils literal notranslate"><span class="pre">adata_embed_with_values()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.alluvial_plot"><code class="docutils literal notranslate"><span class="pre">alluvial_plot()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.diy_cmap_grey_bg"><code class="docutils literal notranslate"><span class="pre">diy_cmap_grey_bg()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.embed_with_values"><code class="docutils literal notranslate"><span class="pre">embed_with_values()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.embedding_mock3d"><code class="docutils literal notranslate"><span class="pre">embedding_mock3d()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.get_colors"><code class="docutils literal notranslate"><span class="pre">get_colors()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.grid_bars_display_probas"><code class="docutils literal notranslate"><span class="pre">grid_bars_display_probas()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.grid_bars_display_probas_transposed"><code class="docutils literal notranslate"><span class="pre">grid_bars_display_probas_transposed()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.grid_display_probas"><code class="docutils literal notranslate"><span class="pre">grid_display_probas()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.heatmap"><code class="docutils literal notranslate"><span class="pre">heatmap()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.heatmap_probas"><code class="docutils literal notranslate"><span class="pre">heatmap_probas()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.multipartite_layout"><code class="docutils literal notranslate"><span class="pre">multipartite_layout()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_confus_mat"><code class="docutils literal notranslate"><span class="pre">plot_confus_mat()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_confus_multi_mats"><code class="docutils literal notranslate"><span class="pre">plot_confus_multi_mats()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_contingency_mat"><code class="docutils literal notranslate"><span class="pre">plot_contingency_mat()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_distance_lines"><code class="docutils literal notranslate"><span class="pre">plot_distance_lines()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_edges_by_adj"><code class="docutils literal notranslate"><span class="pre">plot_edges_by_adj()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_line_list"><code class="docutils literal notranslate"><span class="pre">plot_line_list()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_mapped_graph"><code class="docutils literal notranslate"><span class="pre">plot_mapped_graph()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_multipartite_graph"><code class="docutils literal notranslate"><span class="pre">plot_multipartite_graph()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_records_for_trainer"><code class="docutils literal notranslate"><span class="pre">plot_records_for_trainer()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_splitted_umaps"><code class="docutils literal notranslate"><span class="pre">plot_splitted_umaps()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.plot_stacked_bar"><code class="docutils literal notranslate"><span class="pre">plot_stacked_bar()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.rescale_layout"><code class="docutils literal notranslate"><span class="pre">rescale_layout()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.rotate_xticklabels"><code class="docutils literal notranslate"><span class="pre">rotate_xticklabels()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.rotate_yticklabels"><code class="docutils literal notranslate"><span class="pre">rotate_yticklabels()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.sorted_scatter"><code class="docutils literal notranslate"><span class="pre">sorted_scatter()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.triple_umaps"><code class="docutils literal notranslate"><span class="pre">triple_umaps()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.umap_grid"><code class="docutils literal notranslate"><span class="pre">umap_grid()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.umap_with_annotates"><code class="docutils literal notranslate"><span class="pre">umap_with_annotates()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.venn_plot"><code class="docutils literal notranslate"><span class="pre">venn_plot()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.view_color_map"><code class="docutils literal notranslate"><span class="pre">view_color_map()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.plot.wrapper_heatmap_scores"><code class="docutils literal notranslate"><span class="pre">wrapper_heatmap_scores()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-BrainAlign.came.utils.preprocess">BrainAlign.came.utils.preprocess module</a><ul>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.AdjacentTrans"><code class="docutils literal notranslate"><span class="pre">AdjacentTrans</span></code></a><ul>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.AdjacentTrans.T"><code class="docutils literal notranslate"><span class="pre">AdjacentTrans.T()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.AdjacentTrans.from_edge_df"><code class="docutils literal notranslate"><span class="pre">AdjacentTrans.from_edge_df()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.AdjacentTrans.reduce_to_align"><code class="docutils literal notranslate"><span class="pre">AdjacentTrans.reduce_to_align()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.AdjacentTrans.reduce_to_align_features"><code class="docutils literal notranslate"><span class="pre">AdjacentTrans.reduce_to_align_features()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.AdjacentTrans.shape"><code class="docutils literal notranslate"><span class="pre">AdjacentTrans.shape</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.adata_from_raw"><code class="docutils literal notranslate"><span class="pre">adata_from_raw()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.add_columns"><code class="docutils literal notranslate"><span class="pre">add_columns()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.add_new_edgetype_to_dglgraph"><code class="docutils literal notranslate"><span class="pre">add_new_edgetype_to_dglgraph()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.add_obs_annos"><code class="docutils literal notranslate"><span class="pre">add_obs_annos()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.add_var_annos"><code class="docutils literal notranslate"><span class="pre">add_var_annos()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.agg_group_edges"><code class="docutils literal notranslate"><span class="pre">agg_group_edges()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.align_adata_vars"><code class="docutils literal notranslate"><span class="pre">align_adata_vars()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.all_vars_of_adata"><code class="docutils literal notranslate"><span class="pre">all_vars_of_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.augment_repeat_adata"><code class="docutils literal notranslate"><span class="pre">augment_repeat_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.bisplit_adata"><code class="docutils literal notranslate"><span class="pre">bisplit_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.change_names"><code class="docutils literal notranslate"><span class="pre">change_names()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.compute_and_get_DEGs"><code class="docutils literal notranslate"><span class="pre">compute_and_get_DEGs()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.compute_unkn_rate"><code class="docutils literal notranslate"><span class="pre">compute_unkn_rate()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.compute_unkn_rate_for2"><code class="docutils literal notranslate"><span class="pre">compute_unkn_rate_for2()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.describe_mat"><code class="docutils literal notranslate"><span class="pre">describe_mat()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.describe_mat_nnz"><code class="docutils literal notranslate"><span class="pre">describe_mat_nnz()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.dict_has_keys"><code class="docutils literal notranslate"><span class="pre">dict_has_keys()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.filter_mitogenes"><code class="docutils literal notranslate"><span class="pre">filter_mitogenes()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.get_homologies"><code class="docutils literal notranslate"><span class="pre">get_homologies()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.get_homologues"><code class="docutils literal notranslate"><span class="pre">get_homologues()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.get_hvgs"><code class="docutils literal notranslate"><span class="pre">get_hvgs()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.get_marker_info_table"><code class="docutils literal notranslate"><span class="pre">get_marker_info_table()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.get_marker_name_table"><code class="docutils literal notranslate"><span class="pre">get_marker_name_table()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.get_scnet"><code class="docutils literal notranslate"><span class="pre">get_scnet()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.group_mean"><code class="docutils literal notranslate"><span class="pre">group_mean()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.group_mean_adata"><code class="docutils literal notranslate"><span class="pre">group_mean_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.group_mean_dense"><code class="docutils literal notranslate"><span class="pre">group_mean_dense()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.group_mean_multiadata"><code class="docutils literal notranslate"><span class="pre">group_mean_multiadata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.group_mean_sparse"><code class="docutils literal notranslate"><span class="pre">group_mean_sparse()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.group_median_dense"><code class="docutils literal notranslate"><span class="pre">group_median_dense()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.group_value_counts"><code class="docutils literal notranslate"><span class="pre">group_value_counts()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.group_zscore"><code class="docutils literal notranslate"><span class="pre">group_zscore()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.group_zscore_adata"><code class="docutils literal notranslate"><span class="pre">group_zscore_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.homograph_from_scipy"><code class="docutils literal notranslate"><span class="pre">homograph_from_scipy()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.is1v1pairs"><code class="docutils literal notranslate"><span class="pre">is1v1pairs()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.is_symetric"><code class="docutils literal notranslate"><span class="pre">is_symetric()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.label_binarize_each"><code class="docutils literal notranslate"><span class="pre">label_binarize_each()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.load_dense"><code class="docutils literal notranslate"><span class="pre">load_dense()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.load_namelist"><code class="docutils literal notranslate"><span class="pre">load_namelist()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.load_sparse"><code class="docutils literal notranslate"><span class="pre">load_sparse()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.make_adata"><code class="docutils literal notranslate"><span class="pre">make_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.make_bipartite_adj"><code class="docutils literal notranslate"><span class="pre">make_bipartite_adj()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.make_dict_from_str"><code class="docutils literal notranslate"><span class="pre">make_dict_from_str()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.make_id_name_maps"><code class="docutils literal notranslate"><span class="pre">make_id_name_maps()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.mean_of_nozeros"><code class="docutils literal notranslate"><span class="pre">mean_of_nozeros()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.merge_adata_groups"><code class="docutils literal notranslate"><span class="pre">merge_adata_groups()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.merge_adatas"><code class="docutils literal notranslate"><span class="pre">merge_adatas()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.merge_group_labels"><code class="docutils literal notranslate"><span class="pre">merge_group_labels()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.merge_metas"><code class="docutils literal notranslate"><span class="pre">merge_metas()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.merge_named_matrices"><code class="docutils literal notranslate"><span class="pre">merge_named_matrices()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.mtx2df"><code class="docutils literal notranslate"><span class="pre">mtx2df()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.normalize_col"><code class="docutils literal notranslate"><span class="pre">normalize_col()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.normalize_default"><code class="docutils literal notranslate"><span class="pre">normalize_default()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.normalize_log_then_total"><code class="docutils literal notranslate"><span class="pre">normalize_log_then_total()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.normalize_max"><code class="docutils literal notranslate"><span class="pre">normalize_max()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.normalize_maxmin"><code class="docutils literal notranslate"><span class="pre">normalize_maxmin()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.normalize_norms"><code class="docutils literal notranslate"><span class="pre">normalize_norms()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.normalize_row"><code class="docutils literal notranslate"><span class="pre">normalize_row()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.order_contingency_mat"><code class="docutils literal notranslate"><span class="pre">order_contingency_mat()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.pivot_df_to_sparse"><code class="docutils literal notranslate"><span class="pre">pivot_df_to_sparse()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.pivot_to_sparse"><code class="docutils literal notranslate"><span class="pre">pivot_to_sparse()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.quick_pre_clust"><code class="docutils literal notranslate"><span class="pre">quick_pre_clust()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.quick_pre_vis"><code class="docutils literal notranslate"><span class="pre">quick_pre_vis()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.quick_preprocess"><code class="docutils literal notranslate"><span class="pre">quick_preprocess()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.reduce_to_align"><code class="docutils literal notranslate"><span class="pre">reduce_to_align()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.regu_gname"><code class="docutils literal notranslate"><span class="pre">regu_gname()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.remove_adata_groups"><code class="docutils literal notranslate"><span class="pre">remove_adata_groups()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.remove_adata_small_groups"><code class="docutils literal notranslate"><span class="pre">remove_adata_small_groups()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.remove_small_groups"><code class="docutils literal notranslate"><span class="pre">remove_small_groups()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.reverse_dict"><code class="docutils literal notranslate"><span class="pre">reverse_dict()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.reverse_dict_of_list"><code class="docutils literal notranslate"><span class="pre">reverse_dict_of_list()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.save_mtx2df"><code class="docutils literal notranslate"><span class="pre">save_mtx2df()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.save_named_mtx"><code class="docutils literal notranslate"><span class="pre">save_named_mtx()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.save_namelist"><code class="docutils literal notranslate"><span class="pre">save_namelist()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.scipy_edge_dict_for_dgl"><code class="docutils literal notranslate"><span class="pre">scipy_edge_dict_for_dgl()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.split_adata"><code class="docutils literal notranslate"><span class="pre">split_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.subset_matches"><code class="docutils literal notranslate"><span class="pre">subset_matches()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.take_1v1_matches"><code class="docutils literal notranslate"><span class="pre">take_1v1_matches()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.take_adata_groups"><code class="docutils literal notranslate"><span class="pre">take_adata_groups()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.take_freq1"><code class="docutils literal notranslate"><span class="pre">take_freq1()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.take_freq1more"><code class="docutils literal notranslate"><span class="pre">take_freq1more()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.take_group_labels"><code class="docutils literal notranslate"><span class="pre">take_group_labels()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.top_markers_from_adata"><code class="docutils literal notranslate"><span class="pre">top_markers_from_adata()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.top_markers_from_df"><code class="docutils literal notranslate"><span class="pre">top_markers_from_df()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.top_markers_from_info"><code class="docutils literal notranslate"><span class="pre">top_markers_from_info()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.unpack_dict_of_lists"><code class="docutils literal notranslate"><span class="pre">unpack_dict_of_lists()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.wrapper_normalize"><code class="docutils literal notranslate"><span class="pre">wrapper_normalize()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.wrapper_scale"><code class="docutils literal notranslate"><span class="pre">wrapper_scale()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.preprocess.zscore"><code class="docutils literal notranslate"><span class="pre">zscore()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-BrainAlign.came.utils.train">BrainAlign.came.utils.train module</a><ul>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.Trainer"><code class="docutils literal notranslate"><span class="pre">Trainer</span></code></a><ul>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.Trainer.evaluate_metrics"><code class="docutils literal notranslate"><span class="pre">Trainer.evaluate_metrics()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.Trainer.get_current_outputs"><code class="docutils literal notranslate"><span class="pre">Trainer.get_current_outputs()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.Trainer.log_info"><code class="docutils literal notranslate"><span class="pre">Trainer.log_info()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.Trainer.plot_class_accs"><code class="docutils literal notranslate"><span class="pre">Trainer.plot_class_accs()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.Trainer.plot_class_losses"><code class="docutils literal notranslate"><span class="pre">Trainer.plot_class_losses()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.Trainer.plot_cluster_index"><code class="docutils literal notranslate"><span class="pre">Trainer.plot_cluster_index()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.Trainer.set_class_weights"><code class="docutils literal notranslate"><span class="pre">Trainer.set_class_weights()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.Trainer.train"><code class="docutils literal notranslate"><span class="pre">Trainer.train()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.Trainer.train_minibatch"><code class="docutils literal notranslate"><span class="pre">Trainer.train_minibatch()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.infer_for_nodes"><code class="docutils literal notranslate"><span class="pre">infer_for_nodes()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.make_class_weights"><code class="docutils literal notranslate"><span class="pre">make_class_weights()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.order_by_ids"><code class="docutils literal notranslate"><span class="pre">order_by_ids()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.prepare4train"><code class="docutils literal notranslate"><span class="pre">prepare4train()</span></code></a></li>
<li><a class="reference internal" href="#BrainAlign.came.utils.train.seed_everything"><code class="docutils literal notranslate"><span class="pre">seed_everything()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#brainalign-came-utils-train-v0-module">BrainAlign.came.utils.train_v0 module</a></li>
<li><a class="reference internal" href="#module-BrainAlign.came.utils">Module contents</a></li>
</ul>
</li>
</ul>
</div>
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">BrainAlign</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">BrainAlign.came.utils package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/BrainAlign.came.utils.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="brainalign-came-utils-package">
<h1>BrainAlign.came.utils package<a class="headerlink" href="#brainalign-came-utils-package" title="Permalink to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
</section>
<section id="module-BrainAlign.came.utils.analyze">
<span id="brainalign-came-utils-analyze-module"></span><h2>BrainAlign.came.utils.analyze module<a class="headerlink" href="#module-BrainAlign.came.utils.analyze" title="Permalink to this heading"></a></h2>
<p>Created on Tue Nov  3 13:14:05 2020</p>
<p>&#64;author: Xingyan Liu</p>
<hr class="docutils" />
<p>Functions for downstream biological analysis</p>
<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.abstract_nodes">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">abstract_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_ord</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'size'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_identity</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'identity'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_orig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'original</span> <span class="pre">name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_tag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tag'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.abstract_nodes" title="Permalink to this definition"></a></dt>
<dd><p>Taking each of the groups in <cite>labels</cite> as an abstracted node, and making a
list of node attributes for further input to <cite>networkx.Graph.add_nodes_from()</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key_identity</strong> (<em>key name for the unique identifier for each abstracted node.</em>) – </p></li>
<li><p><strong>key_orig</strong> (<em>key name for the original</em><em> (</em><em>group</em><em>) </em><em>name for each abstracted node.</em>) – </p></li>
<li><p><strong>**kwds</strong> (<em>ignored currently</em>) – </p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df_nodes</span> <span class="o">=</span> <span class="n">abstract_nodes</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;module&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">node_attrs</span><span class="p">,</span> <span class="s1">&#39;node name&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.abstract_ov_edges">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">abstract_ov_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">avg_expr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tag_var</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tag_obs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_first</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_adjust</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_full_adj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.abstract_ov_edges" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>avg_expr</strong> (<em>pd.DataFrame</em>) – each column represent the average expressions
for each observation group, and each row as a variable.</p></li>
<li><p><strong>norm_method</strong> – one of {None, ‘zs’, ‘maxmin’, ‘max’}</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.abstract_vv_edges">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">abstract_vv_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_links:</span> <span class="pre">~pandas.core.frame.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_labels1:</span> <span class="pre">~pandas.core.series.Series</span> <span class="pre">|</span> <span class="pre">~typing.Mapping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_labels2:</span> <span class="pre">~pandas.core.series.Series</span> <span class="pre">|</span> <span class="pre">~typing.Mapping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_sizes:</span> <span class="pre">None</span> <span class="pre">|</span> <span class="pre">~typing.Sequence</span> <span class="pre">|</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_full_adj=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_adjust:</span> <span class="pre">bool</span> <span class="pre">=</span> <span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_norm_func=&lt;built-in</span> <span class="pre">function</span> <span class="pre">max&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_weight:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">'weight'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys_link:</span> <span class="pre">~typing.Sequence[str]</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tag_var1:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tag_var2:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.abstract_vv_edges" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_links</strong> (pd.DataFrame, shape=(n_edges, <a href="#id1"><span class="problematic" id="id2">*</span></a>)) – If <cite>keys_edge</cite> is provided, it should be a tuple of 2 column-names
in df_links.columns, indicating the edge columns.
Otherwise, the <cite>df_links.index</cite> should be <cite>pd.MultiIndex</cite> to indicate
the source and target edges.</p></li>
<li><p><strong>keys_link</strong> – If <cite>keys_edge</cite> is provided, it should be a tuple of 2 column-names
in df_links.columns, indicating the edge columns.</p></li>
<li><p><strong>var_labels1</strong> – grouping labels for the two sets of variables, respectively.</p></li>
<li><p><strong>var_labels2</strong> – grouping labels for the two sets of variables, respectively.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.adata_subgroup_edges">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">adata_subgroup_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'module'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.adata_subgroup_edges" title="Permalink to this definition"></a></dt>
<dd><p>take a subset of the KNN-graph from adata, melt to an edge-dataframe, with
columns [‘source’, ‘target’, ‘weight’] by default.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.adj_to_edges">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">adj_to_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rownames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colnames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_row</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'source'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'target'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'weight'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_attrs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.adj_to_edges" title="Permalink to this definition"></a></dt>
<dd><p>Melt (Un-pivot) a pd.DataFrame of adjacent matrix into a list of edges, works
similar to <cite>pd.melt()</cite> but ignoring ALL of the zeros instead.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.aggregate_links">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">aggregate_links</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_links:</span> <span class="pre">~pandas.core.frame.DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels1:</span> <span class="pre">~pandas.core.series.Series</span> <span class="pre">|</span> <span class="pre">~typing.Mapping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels2:</span> <span class="pre">~pandas.core.series.Series</span> <span class="pre">|</span> <span class="pre">~typing.Mapping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_sizes:</span> <span class="pre">None</span> <span class="pre">|</span> <span class="pre">~typing.Sequence</span> <span class="pre">|</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_adjust:</span> <span class="pre">bool</span> <span class="pre">=</span> <span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_norm_func=&lt;built-in</span> <span class="pre">function</span> <span class="pre">max&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_weight:</span> <span class="pre">str</span> <span class="pre">=</span> <span class="pre">'weight'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys_link:</span> <span class="pre">~typing.Sequence[str]</span> <span class="pre">|</span> <span class="pre">None</span> <span class="pre">=</span> <span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.aggregate_links" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_links</strong> (pd.DataFrame, shape=(n_edges, <a href="#id3"><span class="problematic" id="id4">*</span></a>)) – If <cite>keys_edge</cite> is provided, it should be a tuple of 2 column-names
in df_links.columns, indicating the edge columns.
Otherwise, the <cite>df_links.index</cite> should be <cite>pd.MultiIndex</cite> to indicate
the source and target edges.</p></li>
<li><p><strong>labels1</strong> – grouping labels for the rows and columns, respectively.</p></li>
<li><p><strong>labels2</strong> – grouping labels for the rows and columns, respectively.</p></li>
<li><p><strong>norm_sizes</strong> – if provided, should be a pair of pd.Series, dict, Mapping, list
it can be also set as ‘auto’, which means decide sizes by the group-labels.</p></li>
<li><p><strong>keys_link</strong> (<em>str</em><em> (</em><em>'weight' by default</em><em>)</em>) – If <cite>keys_edge</cite> is provided, it should be a tuple of 2 column-names
in df_links.columns, indicating the edge columns.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.arrange_contingency_mat">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">arrange_contingency_mat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mat</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">novel_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['unknown',</span> <span class="pre">'uncertain',</span> <span class="pre">'multi-type']</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.arrange_contingency_mat" title="Permalink to this definition"></a></dt>
<dd><p>alignment of column and row names</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.compare_deg_dicts">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">compare_deg_dicts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dct_deg1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dct_deg2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.compare_deg_dicts" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dct_deg1</strong> (<em>Dict</em><em>[</em><em>Any</em><em>, </em><em>Sequence</em><em>]</em>) – a dict of DEG lists, where each key corresponds to a cell group.</p></li>
<li><p><strong>dct_deg2</strong> (<em>Dict</em><em>[</em><em>Any</em><em>, </em><em>Sequence</em><em>]</em>) – a dict of DEG lists, where each key corresponds to a cell group.</p></li>
<li><p><strong>df_map</strong> – homologous gene mappings</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>record</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.compare_degs_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">compare_degs_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_padj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_logfc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ntop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_group</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'names'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.compare_degs_adata" title="Permalink to this definition"></a></dt>
<dd><p>Compare DEGs of common cell-types/clusters across datasets (or species)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata1</strong> (<code class="docutils literal notranslate"><span class="pre">sc.AnnData</span></code> objects whose DEGs to be compared) – </p></li>
<li><p><strong>adata2</strong> (<code class="docutils literal notranslate"><span class="pre">sc.AnnData</span></code> objects whose DEGs to be compared) – </p></li>
<li><p><strong>df_map</strong> – homologous gene mappings</p></li>
<li><p><strong>cut_padj</strong> – filter genes with adjusted-p-values lower than this value</p></li>
<li><p><strong>cut_logfc</strong> – filter genes with long-fold-changes higher than this value</p></li>
<li><p><strong>ntop</strong> (<em>Optional</em>) – if specified, take only top-{ntop} DEGs to compare</p></li>
<li><p><strong>key_group</strong> – </p></li>
<li><p><strong>key_gene</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict of dicts</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p># normalize data
&gt;&gt;&gt; pp.normalize_default(adata1)
&gt;&gt;&gt; pp.normalize_default(adata2)
# compute DEGs
&gt;&gt;&gt; sc.tl.rank_genes_groups(adata1, groupby=’cell_ontology_class’)
&gt;&gt;&gt; sc.tl.rank_genes_groups(adata2, groupby=’cell_ontology_class’)
# compute intersections and privates
&gt;&gt;&gt; record = compare_degs_adata(adata1, adata2, df_map, cut_padj=0.05)
# inspect the results
&gt;&gt;&gt; pd.DataFrame(record)
&gt;&gt;&gt; pd.DataFrame(record).applymap(lambda x: len(set(x)))</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.compare_degs_seurat">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">compare_degs_seurat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_deg1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_deg2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_padj</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ntop</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cluster'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_pval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'p_val_adj'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.compare_degs_seurat" title="Permalink to this definition"></a></dt>
<dd><p>Compare DEGs of common cell-types/clusters across datasets (or species)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_deg1</strong> – DEG table output from Seurat’s function FindAllMarkers(), with columns
‘p_val’, ‘avg_logFC’, ‘pct.1’, ‘pct.2’, ‘p_val_adj’, ‘cluster’, ‘gene’</p></li>
<li><p><strong>df_deg2</strong> – DEG table output from Seurat’s function FindAllMarkers(), with the same
format with <code class="docutils literal notranslate"><span class="pre">df_deg1</span></code></p></li>
<li><p><strong>df_map</strong> – homologous gene mappings.
If not given, genes will mapped by their names (case sensitive).</p></li>
<li><p><strong>cut_padj</strong> – filter genes with adjusted-p-values lower than this value</p></li>
<li><p><strong>ntop</strong> (<em>Optional</em>) – if specified, take only top-{ntop} DEGs to compare</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict of dicts</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p># compute intersections and privates
&gt;&gt;&gt; record = compare_degs_adata(df_deg1, df_deg2, df_map, cut_padj=0.05)
# inspect the results
&gt;&gt;&gt; pd.DataFrame(record)
&gt;&gt;&gt; pd.DataFrame(record).applymap(lambda x: len(set(x)))</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.compare_modules">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">compare_modules</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mod_labels1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_labels2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_var_links</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avg_scaled</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zscore_cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Mapping</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Mapping</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#BrainAlign.came.utils.analyze.compare_modules" title="Permalink to this definition"></a></dt>
<dd><p>Compute common and private genes (cross-species) in each gene module.
If <cite>avg_scaled</cite> is provided, the module genes enriched in each cell-type
will be computed and compared.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mod_labels1</strong> (<em>pd.Series</em>) – module labels</p></li>
<li><p><strong>mod_labels2</strong> (<em>pd.Series</em>) – module labels</p></li>
<li><p><strong>df_var_links</strong> (<em>pd.DataFrame</em>) – </p></li>
<li><p><strong>avg_scaled</strong> – If provided, should be a pair of DataFrame storing the average
expressions for each dataset (species), and the index should be
the gene names.</p></li>
<li><p><strong>zscore_cut</strong> – Cut-off of expression z-scores. This will be ignored if <code class="docutils literal notranslate"><span class="pre">avg_scaled</span></code>
is not provided.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>record</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict of dicts</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.compute_common_private">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">compute_common_private</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gmap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.compute_common_private" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>compute common and private genes based on a given gene mapping</dt><dd><p>(e.g.,homologous mapping)</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genes1</strong> – the two gene sets to compare</p></li>
<li><p><strong>genes2</strong> – the two gene sets to compare</p></li>
<li><p><strong>gmap</strong> – a DataFrame with at least two columns, storing homologous gene mappings.
the first column corresponds to <code class="docutils literal notranslate"><span class="pre">genes1</span></code>, and the second corresponds
to <code class="docutils literal notranslate"><span class="pre">genes2</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>record</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.edges_from_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">edges_from_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kind</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'conn'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_neigh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_row</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'source'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'target'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_attrs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.edges_from_adata" title="Permalink to this definition"></a></dt>
<dd><p>transform the connectivities in <cite>adata</cite> into an edge-dataframe, with
columns [‘source’, ‘target’, ‘weight’] by default.</p>
<p>kind: ‘conn’, ‘dist’</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.export_neighbor_subgraph_df">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">export_neighbor_subgraph_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodenames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">WindowsPath('.')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tfset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.export_neighbor_subgraph_df" title="Permalink to this definition"></a></dt>
<dd><p>temp function;</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.export_subgraph_df">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">export_subgraph_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodenames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">WindowsPath('.')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tfset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.export_subgraph_df" title="Permalink to this definition"></a></dt>
<dd><p>export graph to ‘.csv’ files for other visualization</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.get_adata_neighbors">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">get_adata_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.get_adata_neighbors" title="Permalink to this definition"></a></dt>
<dd><p>getting distances and connectivities from adata</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.load_dpair_and_model">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">load_dpair_and_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subdir_model</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'_models'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ckpt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.load_dpair_and_model" title="Permalink to this definition"></a></dt>
<dd><p>Load the output results of CAME.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dirname</strong> – result directory of CAME</p></li>
<li><p><strong>subdir_model</strong> – subdirectory where the model checkpoints are saved.</p></li>
<li><p><strong>ckpt</strong> – specify which model checkpoint to load.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>dpair</strong> (<em>AlignedDataPair or DataPair</em>) – the datapair object that stores the features and graph</p></li>
<li><p><strong>model</strong> (<em>CGGCNet or CGCNet</em>) – the graph neural network model</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dpair</span><span class="p">,</span> <span class="n">model</span> <span class="o">=</span> <span class="n">came</span><span class="o">.</span><span class="n">load_dpair_and_model</span><span class="p">(</span><span class="n">came_resdir</span><span class="p">)</span>
</pre></div>
</div>
<p>access the feature dict</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">feat_dict</span> <span class="o">=</span> <span class="n">dpair</span><span class="o">.</span><span class="n">get_feature_dict</span><span class="p">(</span><span class="n">scale</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>access the heterogrnrous cell-gene graph</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">dpair</span><span class="o">.</span><span class="n">get_whole_net</span><span class="p">()</span>
</pre></div>
</div>
<p>access the reference and query sample-ids</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">obs_ids1</span><span class="p">,</span> <span class="n">obs_ids2</span> <span class="o">=</span> <span class="n">dpair</span><span class="o">.</span><span class="n">obs_ids1</span><span class="p">,</span> <span class="n">dpair</span><span class="o">.</span><span class="n">obs_ids2</span>
</pre></div>
</div>
<p>passing forward</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">outputs</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">feat_dict</span><span class="p">,</span> <span class="n">g</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.make_abstracted_graph">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">make_abstracted_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obs_labels1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_labels2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_labels1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_labels2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avg_expr1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avg_expr2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_var_links</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags_obs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('',</span> <span class="pre">'')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags_var</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('',</span> <span class="pre">'')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_weight</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'weight'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_ov</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_mtd_ov</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'zs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ov_norm_first</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_adjust_ov</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_adjust_vv</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vargroup_filtered</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'filtered'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.make_abstracted_graph" title="Permalink to this definition"></a></dt>
<dd><p>Compute and make the abstracted graph from expression matrices and the
linkage weights between homologous genes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>obs_labels1</strong> – group labels of the reference and query observations (cells), respectively.</p></li>
<li><p><strong>obs_labels2</strong> – group labels of the reference and query observations (cells), respectively.</p></li>
<li><p><strong>var_labels1</strong> – group labels of the reference and query variables (genes), respectively.</p></li>
<li><p><strong>var_labels2</strong> – group labels of the reference and query variables (genes), respectively.</p></li>
<li><p><strong>avg_expr1</strong> – averaged expression matrix of the reference data</p></li>
<li><p><strong>avg_expr2</strong> – averaged expression matrix of the query data</p></li>
<li><p><strong>df_var_links</strong> – the linkage-weights between homologous genes</p></li>
<li><p><strong>tags_obs</strong> – a tuple of two strings, for specifying homologous cel-types from
different species.
For example, if set <code class="docutils literal notranslate"><span class="pre">tags_obs=('human</span> <span class="pre">',</span> <span class="pre">'mouse</span> <span class="pre">')</span></code>, the result
node names (for ‘T cell’) will be ‘human T cell’ and ‘mouse T cell’,
respectively.</p></li>
<li><p><strong>tags_var</strong> – a tuple of two strings, for specifying gene modules from different
species.
For example, if set <code class="docutils literal notranslate"><span class="pre">tags_obs=('human</span> <span class="pre">module</span> <span class="pre">',</span> <span class="pre">'mouse</span> <span class="pre">module</span> <span class="pre">')</span></code>,
the result node names (for gene module ‘1’) will be ‘human module 1’
and ‘mouse module 1’, respectively.</p></li>
<li><p><strong>key_weight</strong> – column name in <code class="docutils literal notranslate"><span class="pre">df_var_links</span></code>, specifying weights between each pair
of variables.</p></li>
<li><p><strong>cut_ov</strong> – the threshold to cut edges with values lower than it.</p></li>
<li><p><strong>norm_mtd_ov</strong> – one of {None, ‘zs’, ‘maxmin’, ‘max’}</p></li>
<li><p><strong>global_adjust_ov</strong> – whether to globally adjust the weights between the observations and
the variables</p></li>
<li><p><strong>global_adjust_vv</strong> – whether to globally adjust the weights between the observations</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.make_nx_input_from_df">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">make_nx_input_from_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.make_nx_input_from_df" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame with each column containing node attributes</em>) – </p></li>
<li><p><strong>key_id</strong> – which column(s) should be used for unique identifier for each nodes.
str or list of strings from df.columns, or None (using <cite>df.index</cite> in this case)
If specified, the values should be unique.</p></li>
<li><p><strong>**attrs</strong> – other common attributes for this batch of nodes</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>A list of tuples, with each tuple formed like <cite>(node, attrdict)</cite> when</p></li>
<li><p><cite>len(key_id) == 1</cite> or <cite>(u, v, attrdict)</cite> when <cite>len(key_id) == 2</cite></p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">node_attrs</span> <span class="o">=</span> <span class="n">make_nx_input_from_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">**</span><span class="n">attrs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">add_nodes_from</span><span class="p">(</span><span class="n">node_attrs</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.module_enrichment_for_classes">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">module_enrichment_for_classes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">avg_scaled1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avg_scaled2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_genes1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_genes2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes_common1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes_common2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zscore_cut</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">ignored</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.module_enrichment_for_classes" title="Permalink to this definition"></a></dt>
<dd><p>For each (cell-)type and the given gene set (module) of two species,
calculate the relatively highly expressed genes, and find those genes that
are highly expressed in both species, and species-specific gene.</p>
<p>Note that <cite>genes_common1</cite> and <cite>genes_common2</cite> should be of the same length.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>record</strong></p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict of dicts</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.module_homo_weights">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">module_homo_weights</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">var</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_links</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">module</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_module</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'module'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_split</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dataset'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_private</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.module_homo_weights" title="Permalink to this definition"></a></dt>
<dd><p>extract the weights between homologous genes in the given module</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>var</strong> – a DataFrame or a pair of DataFrames storing the annotations of variables
(genes), should be with columns {<cite>key_module</cite>, <cite>key_split</cite>, <cite>key_name</cite>}</p></li>
<li><p><strong>df_links</strong> – a DataFrame recording the weights between each pair of homologies.</p></li>
<li><p><strong>module</strong> – module name, e.g. ‘0’, ‘1’, …</p></li>
<li><p><strong>key_module</strong> – column name in var, of which the module-identities are stored</p></li>
<li><p><strong>key_split</strong> – a column name in var, to split by (if var is a single DataFrame).</p></li>
<li><p><strong>key_name</strong> – a column name in var, storing gene names</p></li>
<li><p><strong>include_private</strong> – whether to include the weights (filled by zeros) between
dataset/species-specific genes</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>a list of weights between homologous genes in the given module.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.neighbor_induced_edgedf">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">neighbor_induced_edgedf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys_edge</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['source',</span> <span class="pre">'target']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.neighbor_induced_edgedf" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.neighbor_induced_subgraph">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">neighbor_induced_subgraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodes</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.neighbor_induced_subgraph" title="Permalink to this definition"></a></dt>
<dd><p>Note that the returned graph also contains the edges connecting the
neighbor-nodes if there is any.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.nx_from_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">nx_from_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_neigh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Graph</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.analyze.nx_from_adata" title="Permalink to this definition"></a></dt>
<dd><p>nx.Graph from the KNN graph of <cite>adata</cite></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.nx_multipartite_graph">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">nx_multipartite_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">node_layers</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'subset'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">attrs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.nx_multipartite_graph" title="Permalink to this definition"></a></dt>
<dd><p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">nx_multipartite_graph</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span> <span class="p">[</span><span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">multipartite_layout</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">subset_key</span><span class="o">=</span><span class="n">subset_key</span><span class="p">,</span> <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nx</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">with_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.nx_neighbors">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">nx_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_self</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.nx_neighbors" title="Permalink to this definition"></a></dt>
<dd><p>find all the neighbors of given nodes</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.nx_to_dfs">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">nx_to_dfs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodelist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'source'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'target'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.nx_to_dfs" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>edgedf, nodedf</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">edgedf</span><span class="p">,</span> <span class="n">nodedf</span> <span class="o">=</span> <span class="n">nx_to_dfs</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.nx_to_nodedf">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">nx_to_nodedf</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodelist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.nx_to_nodedf" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.set_precomputed_neighbors">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">set_precomputed_neighbors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">distances</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">connectivities</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_neighbors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cosine'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric_kwds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_rep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pcs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.set_precomputed_neighbors" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.weight_linked_vars">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">weight_linked_vars</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">spmatrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cosine'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">func_dist2weight</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Callable</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sigma</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sort</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">1)</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">ignored</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.analyze.weight_linked_vars" title="Permalink to this definition"></a></dt>
<dd><p>Computes the similarity of each linked (homologous) pair of variables.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>np.ndarray</em>) – feature matrix of shape (N, M), where N is the number of sample and M is
the feature dimensionality.</p></li>
<li><p><strong>adj</strong> – sparse.spmatrix; binary adjacent matrix of shape (N, N). Note that only
the upper triangle of the matrix will be considered!</p></li>
<li><p><strong>names</strong> – a sequence of names for rows of <cite>X</cite>, of shape (N,)</p></li>
<li><p><strong>metric</strong> – the metric to quantify the similarities of the given vectors (embeddings)</p></li>
<li><p><strong>sort</strong> – whether to sort by the resulting weights</p></li>
<li><p><strong>index_names</strong> – a pair of names for the multi-index of the resulting DataFrame.
e.g., a pair of dataset or species names (in cross-species scenario)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>df</strong> – with columns [<code class="docutils literal notranslate"><span class="pre">index_names[0]</span></code>, <code class="docutils literal notranslate"><span class="pre">index_names[1]</span></code>, “distance”, “weight”]</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.weight_linked_vars_by_expr">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">weight_linked_vars_by_expr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dpair</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="BrainAlign.came.datapair.html#BrainAlign.came.datapair.unaligned.DataPair" title="BrainAlign.came.datapair.unaligned.DataPair"><span class="pre">DataPair</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels_or_probs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.weight_linked_vars_by_expr" title="Permalink to this definition"></a></dt>
<dd><p>Compute the weights between homologies by their average expressions
across (cell) groups.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dpair</strong> – the <code class="docutils literal notranslate"><span class="pre">DataPair</span></code> object</p></li>
<li><p><strong>labels_or_probs</strong> – group (cell-type) labels or soft assignments (probabilities)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>df_var_links_expr</strong> (<em>pd.DataFrame</em>) – the weights between each pair of homologous genes</p></li>
<li><p><strong>avg</strong> (<em>pd.DataFrame</em>) – concatenated average expressions</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.weight_normalize_by_size">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">weight_normalize_by_size</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sizes1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sizes2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_func=&lt;built-in</span> <span class="pre">function</span> <span class="pre">max&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">global_adjust=False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.weight_normalize_by_size" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adj</strong> (<em>pd.DataFrame</em>) – adjacent matrix of shape (n1, n2)</p></li>
<li><p><strong>sizes1</strong> (<em>pd.Series</em><em>, </em><em>dict</em><em>, </em><em>Mapping</em><em>, </em><em>list-like</em>) – shape = (n1,) and (n2,)</p></li>
<li><p><strong>sizes2</strong> (<em>pd.Series</em><em>, </em><em>dict</em><em>, </em><em>Mapping</em><em>, </em><em>list-like</em>) – shape = (n1,) and (n2,)</p></li>
<li><p><strong>global_adjust</strong> (<em>bool</em>) – whether to perform a global adjustment after
size-normalization.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.wrapper_confus_mat">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">wrapper_confus_mat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y_true</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_pred</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes_on</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'true'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.wrapper_confus_mat" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>normalize: ‘true’, ‘pred’, ‘all’, None</dt><dd><p>by default, normalized by row (true classes)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.analyze.wrapper_contingency_mat">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.analyze.</span></span><span class="sig-name descname"><span class="pre">wrapper_contingency_mat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y_true</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_pred</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_rows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assparse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.analyze.wrapper_contingency_mat" title="Permalink to this definition"></a></dt>
<dd><p>Modified and wrapped function from <cite>sklearn</cite>:
&gt;&gt;&gt; mat = sklearn.metrics.cluster.contingency_matrix(
…        y_true, y_pred, eps=eps, sparse=assparse)</p>
</dd></dl>

</section>
<section id="module-BrainAlign.came.utils.base">
<span id="brainalign-came-utils-base-module"></span><h2>BrainAlign.came.utils.base module<a class="headerlink" href="#module-BrainAlign.came.utils.base" title="Permalink to this heading"></a></h2>
<p>Created on Sat Feb 20 12:16:32 2021</p>
<p>&#64;author: Xingyan Liu</p>
<blockquote>
<div><p>Basic functions</p>
</div></blockquote>
<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.check_dirs">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">check_dirs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.check_dirs" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.dec_timewrapper">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">dec_timewrapper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'function'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.dec_timewrapper" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.dict_struct">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">dict_struct</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'|-'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.dict_struct" title="Permalink to this definition"></a></dt>
<dd><p>visualize the structure of a dict</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.dict_to_pairs">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">dict_to_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dct</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.dict_to_pairs" title="Permalink to this definition"></a></dt>
<dd><p>pairs = []
for k, lst in dct.items():</p>
<blockquote>
<div><p>pairs.extend([(k, v) for v in lst])</p>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.load_json_dict">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">load_json_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf-8'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.load_json_dict" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.load_pickle">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">load_pickle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fp</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.load_pickle" title="Permalink to this definition"></a></dt>
<dd><p>load the object from a .pickle file</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">load_pickle</span><span class="p">(</span><span class="s1">&#39;dpair.pickle&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.make_nowtime_tag">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">make_nowtime_tag</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nowtime</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">brackets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.make_nowtime_tag" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.make_pairs_from_lists">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">make_pairs_from_lists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lst2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inverse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_equal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.make_pairs_from_lists" title="Permalink to this definition"></a></dt>
<dd><p>making combinational pairs</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.map_by_sme">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">map_by_sme</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pairs_sm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pairs_me</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_pairs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.base.map_by_sme" title="Permalink to this definition"></a></dt>
<dd><p>start - media - end homologous mapping transform</p>
<dl class="simple">
<dt>pairs_sm, pairs_me:</dt><dd><p>a list of tuples</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dct_se</strong></p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>a dict of lists</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>make pairs (List[Tuple]) from dfs, mapping by a media
&gt;&gt;&gt; pairs_sm = df1.iloc[:, :2].apply(tuple, axis=1).tolist()
&gt;&gt;&gt; pairs_me = df2.iloc[:, :2].apply(tuple, axis=1).tolist()
&gt;&gt;&gt; pairs_se = map_by_sme(pairs_sm, pairs_me, to_pairs=True)
&gt;&gt;&gt; pd.DataFrame(pairs_se, columns=[‘amph_id’, ‘zebrafish_id’])</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.pairs_to_dict">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">pairs_to_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pairs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.pairs_to_dict" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>pairs:</dt><dd><p>a list of tuples</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dct</strong></p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>a dict of lists</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.save_json_dict">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">save_json_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dct</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'test_json.json'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">encoding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'utf-8'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.save_json_dict" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.save_pickle">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">save_pickle</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">obj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.save_pickle" title="Permalink to this definition"></a></dt>
<dd><p>save the object into a .pickle file</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.split_df">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">split_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.split_df" title="Permalink to this definition"></a></dt>
<dd><p>Split a DataFrame into multiple ones by the given column</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – the DataFrame to split</p></li>
<li><p><strong>by</strong> – a column name in df, to group and split by.</p></li>
<li><p><strong>var1</strong> (<em>&gt;&gt;&gt;</em>) – </p></li>
<li><p><strong>split_df</strong><strong>(</strong><strong>var</strong> (<em>var2 =</em>) – </p></li>
<li><p><strong>'dataset'</strong><strong>)</strong> – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.strline">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">strline</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.strline" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.subsample_each_group">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">subsample_each_group</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">group_labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.subsample_each_group" title="Permalink to this definition"></a></dt>
<dd><p>randomly sample indices from each group, labeled by <cite>group_labels</cite>.
and return the sampled indices (original-order-kept).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>group_labels</strong> – group labels of each point</p></li>
<li><p><strong>n_out</strong> (<em>int</em>) – number of samples for each group</p></li>
<li><p><strong>seed</strong> – the random seed</p></li>
<li><p><strong>groups</strong> – groups to be subsampled and returned.
If None, all groups in <cite>group_labels</cite> will be subsampled</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.subsample_single">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">subsample_single</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">N</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.subsample_single" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>N:</dt><dd><p>The total number of the original indices to be subsampled</p>
</dd>
<dt>frac:</dt><dd><p>Subsample to this <cite>fraction</cite> of the number of indices.</p>
</dd>
<dt>n_min:</dt><dd><p>If the resulting number of indices is less than <cite>n_min</cite>, then
<cite>min([n_min, N])</cite> indices will be sampled, where <cite>N</cite> is
the total number of the original indices.</p>
</dd>
<dt>n_out:</dt><dd><p>if specified, it must be smaller than N, and the two arguments
<cite>frac</cite> and <cite>n_min</cite> will be ignored.</p>
</dd>
<dt>seed:</dt><dd><p>random state</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.base.write_info">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.base.</span></span><span class="sig-name descname"><span class="pre">write_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fn</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">dicts</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.base.write_info" title="Permalink to this definition"></a></dt>
<dd><p>key words parameter-dicts</p>
</dd></dl>

</section>
<section id="module-BrainAlign.came.utils.downsample_counts">
<span id="brainalign-came-utils-downsample-counts-module"></span><h2>BrainAlign.came.utils.downsample_counts module<a class="headerlink" href="#module-BrainAlign.came.utils.downsample_counts" title="Permalink to this heading"></a></h2>
<p>&#64;author: Xingyan Liu
&#64;file: vis_hidden_layers.py
&#64;time: Dec 23 22:06:09 2020</p>
<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.downsample_counts.downsample_counts">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.downsample_counts.</span></span><span class="sig-name descname"><span class="pre">downsample_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">counts_per_cell</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Collection</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">total_counts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.downsample_counts.downsample_counts" title="Permalink to this definition"></a></dt>
<dd><p>Downsample counts from count matrix.</p>
<p>If <cite>counts_per_cell</cite> is specified, each cell will downsampled.
If <cite>total_counts</cite> is specified, expression matrix will be downsampled to
contain at most <cite>total_counts</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Annotated data matrix.</p></li>
<li><p><strong>counts_per_cell</strong> – Target total counts per cell. If a cell has more than ‘counts_per_cell’,
it will be downsampled to this number. Resulting counts can be specified
on a per cell basis by passing an array.Should be an integer or integer
ndarray with same length as number of obs.</p></li>
<li><p><strong>total_counts</strong> – Target total counts. If the count matrix has more than <cite>total_counts</cite>
it will be downsampled to have this number.</p></li>
<li><p><strong>random_state</strong> – Random seed for subsampling.</p></li>
<li><p><strong>replace</strong> – Whether to sample the counts with replacement.</p></li>
<li><p><strong>copy</strong> – Determines whether a copy of <cite>adata</cite> is returned.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Depending on <cite>copy</cite> returns or updates an <cite>adata</cite> with downsampled <cite>.X</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.downsample_counts.downsample_counts_per_cell">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.downsample_counts.</span></span><span class="sig-name descname"><span class="pre">downsample_counts_per_cell</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.downsample_counts.downsample_counts_per_cell" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.downsample_counts.downsample_total_counts">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.downsample_counts.</span></span><span class="sig-name descname"><span class="pre">downsample_total_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frac</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.downsample_counts.downsample_total_counts" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-BrainAlign.came.utils.evaluation">
<span id="brainalign-came-utils-evaluation-module"></span><h2>BrainAlign.came.utils.evaluation module<a class="headerlink" href="#module-BrainAlign.came.utils.evaluation" title="Permalink to this heading"></a></h2>
<p>Created on Sun Apr 11 19:43:10 2021</p>
<p>&#64;author: Xingyan Liu</p>
<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.evaluation.accuracy">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.evaluation.</span></span><span class="sig-name descname"><span class="pre">accuracy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">logits</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tensor</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.evaluation.accuracy" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.evaluation.get_AMI">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.evaluation.</span></span><span class="sig-name descname"><span class="pre">get_AMI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y_true</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_pred</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.evaluation.get_AMI" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.evaluation.get_F1_score">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.evaluation.</span></span><span class="sig-name descname"><span class="pre">get_F1_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y_true</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_pred</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">average</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'macro'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.evaluation.get_F1_score" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-BrainAlign.came.utils.plot">
<span id="brainalign-came-utils-plot-module"></span><h2>BrainAlign.came.utils.plot module<a class="headerlink" href="#module-BrainAlign.came.utils.plot" title="Permalink to this heading"></a></h2>
<p>Created on Fri Oct 23 12:31:28 2020</p>
<p>&#64;author: Xingyan Liu</p>
<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.adata_embed_with_values">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">adata_embed_with_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Mapping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">embed_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'UMAP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize_save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.adata_embed_with_values" title="Permalink to this definition"></a></dt>
<dd><p>This function can also be used to visualize the average expressions of some
cell types on gene embedding, in which case, each observation in
<cite>adata</cite> represents a genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object</p></li>
<li><p><strong>values</strong> (<em>Mapping</em><em>, </em><em>pd.DataFrame</em><em> or </em><em>pd.Series</em>) – where the keys will be the title of subplots and the values
should be of the length equal to adata.shape[0]</p></li>
<li><p><strong>embed_key</strong> – the type of the embeddings, e.g., “UMAP” or “TSNE”</p></li>
<li><p><strong>fp</strong> (<em>Path</em><em> or </em><em>str</em>) – path to save figure</p></li>
<li><p><strong>figsize_save</strong> (<em>(</em><em>int</em><em>, </em><em>int</em><em>)</em>) – the figure size of the saved plot.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>fig, axs</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.alluvial_plot">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">alluvial_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">confsdf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['',</span> <span class="pre">'']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">label_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'alluvial</span> <span class="pre">plot'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.75</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tab20_r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shuffle_colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fonsize_title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">11</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">a_sort</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">b_sort</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.alluvial_plot" title="Permalink to this definition"></a></dt>
<dd><p>visualizing confusion matrix using alluvial plot</p>
<p>a_sort, b_sort: sorted group names on the left and the right side</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.diy_cmap_grey_bg">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">diy_cmap_grey_bg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name_fg</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RdPu'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">low</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rm_high</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.diy_cmap_grey_bg" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.embed_with_values">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">embed_with_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Series</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Mapping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">do_zscore</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_xy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'UMAP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RdYlBu_r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.embed_with_values" title="Permalink to this definition"></a></dt>
<dd><p>Visualize the given x-y coordinates, colored by the given (dict of) values</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>xy</strong> (<em>np.array</em>) – coordinates of each points, of shape (n_points, 2)</p></li>
<li><p><strong>values</strong> (<em>Mapping</em><em>, </em><em>pd.DataFrame</em><em> or </em><em>pd.Series</em>) – where the keys will be the title of subplots and each of the values
should be a vector of the length equal to xy.shape[0]</p></li>
<li><p><strong>do_zscore</strong> – whether to calculate and display the z-scores instead of the original
values. (False)</p></li>
<li><p><strong>s</strong> – the size of the scatters / points (1.5)</p></li>
<li><p><strong>ncols</strong> – the number of columns for multiple-subplots (5)</p></li>
<li><p><strong>axscale</strong> – adjust the size of the figures (2.5)</p></li>
<li><p><strong>with_cbar</strong> – whether to display the color bars (True)</p></li>
<li><p><strong>cmap</strong> – the colormap</p></li>
<li><p><strong>vmin</strong> – the minimum and the maximum of the values</p></li>
<li><p><strong>vmax</strong> – the minimum and the maximum of the values</p></li>
<li><p><strong>kwargs</strong> – Other Parameters for <code class="docutils literal notranslate"><span class="pre">sorted_scatter</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.embedding_mock3d">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">embedding_mock3d</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'module'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edges</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">neighbor_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pt_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">angle_x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">angle_y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_pt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tab20'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(4,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds_edge</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.embedding_mock3d" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – </p></li>
<li><p><strong>color</strong> – a key in adata.obs</p></li>
<li><p><strong>edges</strong> – </p></li>
<li><p><strong>neighbor_key</strong> – </p></li>
<li><p><strong>pt_size</strong> – </p></li>
<li><p><strong>angle_x</strong> – </p></li>
<li><p><strong>angle_y</strong> – </p></li>
<li><p><strong>cmap_pt</strong> – </p></li>
<li><p><strong>ax</strong> – </p></li>
<li><p><strong>figsize</strong> – </p></li>
<li><p><strong>fp</strong> – </p></li>
<li><p><strong>kwds_edge</strong> – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.get_colors">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">get_colors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">'Reds'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.get_colors" title="Permalink to this definition"></a></dt>
<dd><p>fx.view_color_map(‘Reds’, 4)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.grid_bars_display_probas">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">grid_bars_display_probas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(6,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sharey</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.25,</span> <span class="pre">1.05)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.grid_bars_display_probas" title="Permalink to this definition"></a></dt>
<dd><p>bar plots of the distributions</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.grid_bars_display_probas_transposed">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">grid_bars_display_probas_transposed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sharey</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.25,</span> <span class="pre">1.05)</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.grid_bars_display_probas_transposed" title="Permalink to this definition"></a></dt>
<dd><p>bar plots of the distributions (transposed)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.grid_display_probas">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">grid_display_probas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(6,</span> <span class="pre">5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sharey</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(-0.1,</span> <span class="pre">1.1)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.grid_display_probas" title="Permalink to this definition"></a></dt>
<dd><p>violin plots of the distributions</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.heatmap">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_hmap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_row</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(6,</span> <span class="pre">6)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'magma_r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbar_kws</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'shrink':</span> <span class="pre">0.5}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xrotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">45</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">yrotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.heatmap" title="Permalink to this definition"></a></dt>
<dd><p>wrapper of sns.heatmap()</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>norm_method</strong> (<em>{None</em><em>, </em><em>'zs'</em><em>, </em><em>'max'</em><em>, </em><em>'maxmin'}</em>) – normalization method</p></li>
<li><p><strong>norm_axis</strong> (<em>0</em><em> or </em><em>1</em>) – 0 for column-normalization and 1 for rows.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em>) – (6, 6) by default, ignored if ax is not None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.heatmap_probas">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">heatmap_probas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lbs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_heat</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'magma_r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_lb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tab20'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xrotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.heatmap_probas" title="Permalink to this definition"></a></dt>
<dd><p>Heatmap of the prediction probabilities</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_data</strong> – A DataFrame of shape <code class="docutils literal notranslate"><span class="pre">(n_samples,</span> <span class="pre">n_ref_classes)</span></code>.
Note that the samples should be sorted by the (query) group labels,
i.e., <code class="docutils literal notranslate"><span class="pre">lbs</span></code>.</p></li>
<li><p><strong>lbs</strong> – the (query) group labels</p></li>
<li><p><strong>name_label</strong> – the name of the group labels</p></li>
<li><p><strong>cmap_heat</strong> – the color-map of the heatmap</p></li>
<li><p><strong>cmap_lb</strong> – the color-map of the group labels (<code class="docutils literal notranslate"><span class="pre">lbs</span></code>).</p></li>
<li><p><strong>figsize</strong> – the figure size</p></li>
<li><p><strong>vmax</strong> – the maximum and the minimum of the heatmap</p></li>
<li><p><strong>vmin</strong> – the maximum and the minimum of the heatmap</p></li>
<li><p><strong>xrotation</strong> – rotation of the x-tick-labels</p></li>
<li><p><strong>fp</strong> – file-path for saving the plot</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>gs (gridspec)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.multipartite_layout">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">multipartite_layout</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">G</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'subset'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'vertical'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spread_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.multipartite_layout" title="Permalink to this definition"></a></dt>
<dd><p>Position nodes in layers of straight lines.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>G</strong> (<em>NetworkX graph</em><em> or </em><em>list</em><em> of </em><em>nodes</em>) – A position will be assigned to every node in G.</p></li>
<li><p><strong>subset_key</strong> (<em>string</em><em> (</em><em>default='subset'</em><em>)</em>) – Key of node data to be used as layer subset.</p></li>
<li><p><strong>align</strong> (<em>string</em><em> (</em><em>default='vertical'</em><em>)</em>) – The alignment of nodes. Vertical or horizontal.</p></li>
<li><p><strong>scale</strong> (<em>number</em><em> (</em><em>default: 1</em><em>)</em>) – Scale factor for positions.</p></li>
<li><p><strong>center</strong> (<em>array-like</em><em> or </em><em>None</em>) – Coordinate pair around which to center the layout.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pos</strong> – A dictionary of positions keyed by node.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">G</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">complete_multipartite_graph</span><span class="p">(</span><span class="mi">28</span><span class="p">,</span> <span class="mi">16</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pos</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">multipartite_layout</span><span class="p">(</span><span class="n">G</span><span class="p">)</span>
</pre></div>
</div>
<p class="rubric">Notes</p>
<p>This algorithm currently only works in two dimensions and does not
try to minimize edge crossings.</p>
<p>Network does not need to be a complete multipartite graph. As long as nodes
have subset_key data, they will be placed in the corresponding layers.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_confus_mat">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_confus_mat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y_true</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_pred</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes_on</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'true'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linewidths</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.02</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">linecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'grey'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(4,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_confus_mat" title="Permalink to this definition"></a></dt>
<dd><p>function for plotting the contingency matrix
by default, the values will be normalized by row (true classes)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_confus_multi_mats">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_confus_multi_mats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ytrue_lists</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ypred_lists</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes_on</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nrows</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(8,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_confus_multi_mats" title="Permalink to this definition"></a></dt>
<dd><p>combime multiple confusion matrix-plots into a single plot</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_contingency_mat">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_contingency_mat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y_true</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_pred</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">arrange</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(4,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_contingency_mat" title="Permalink to this definition"></a></dt>
<dd><p>function for plotting the contingency matrix</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_distance_lines">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_distance_lines</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dists</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'distance'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">markersize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_distance_lines" title="Permalink to this definition"></a></dt>
<dd><p>names: a list of str, yticklabels</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_edges_by_adj">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_edges_by_adj</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adj</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">spmatrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds_edge</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_edges_by_adj" title="Permalink to this definition"></a></dt>
<dd><p>adj: a sparse matrix, if None, nothing will be plotted</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_line_list">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_line_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ys</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lbs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(4.5,</span> <span class="pre">3.5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_loc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1.05,</span> <span class="pre">0)</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_line_list" title="Permalink to this definition"></a></dt>
<dd><p>ys: a list of lists, each sub-list is a set of curve-points to be plotted.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_mapped_graph">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_mapped_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">xy1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xy2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mapping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pt_color1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pt_color2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap_pt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'tab20'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">angle_x</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">angle_y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offset_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">yscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pt_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">style_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'default'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(5,</span> <span class="pre">7)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds_edge</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_mapped_graph" title="Permalink to this definition"></a></dt>
<dd><p>pt_color1, pt_color2: RGB colors or numerical label sequence</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_multipartite_graph">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_multipartite_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'subset'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'weight'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodelb_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'original</span> <span class="pre">name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.85</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(10,</span> <span class="pre">8)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">yscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_size_min</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_size_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1800</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_multipartite_graph" title="Permalink to this definition"></a></dt>
<dd><p>display the abstracted multipartite graph
(cellType - geneModules - geneModules - cellTypes)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_records_for_trainer">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_records_for_trainer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">trainer</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">record_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lbs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'training</span> <span class="pre">logs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_records_for_trainer" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_splitted_umaps">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_splitted_umaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">splitby</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(4,</span> <span class="pre">8)</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_splitted_umaps" title="Permalink to this definition"></a></dt>
<dd><p>multiple-plots of embeddings</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.plot_stacked_bar">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">plot_stacked_bar</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(6,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Callable</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'tab20b'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_loc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1.02,</span> <span class="pre">0.01)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.plot_stacked_bar" title="Permalink to this definition"></a></dt>
<dd><p>helper function for visualizing the group compositions (e.g., in
each stage or condition).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – the values in each row will be display as one stacked bar.</p></li>
<li><p><strong>norm</strong> (<em>bool</em>) – whether to normalize each row to unit-sum.</p></li>
<li><p><strong>figsize</strong> (<em>tuple</em>) – </p></li>
<li><p><strong>colors</strong> (<em>Sequence</em>) – </p></li>
<li><p><strong>cmap</strong> (<em>Union</em><em>[</em><em>str</em><em>, </em><em>Callable</em><em>]</em>) – color-map function or color-map name</p></li>
<li><p><strong>legend_loc</strong> (<em>tuple</em>) – the legend location</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>ax</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.rescale_layout">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">rescale_layout</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.rescale_layout" title="Permalink to this definition"></a></dt>
<dd><p>Returns scaled position array to (-scale, scale) in all axes.</p>
<p>The function acts on NumPy arrays which hold position information.
Each position is one row of the array. The dimension of the space
equals the number of columns. Each coordinate in one column.</p>
<p>To rescale, the mean (center) is subtracted from each axis separately.
Then all values are scaled so that the largest magnitude value
from all axes equals <cite>scale</cite> (thus, the aspect ratio is preserved).
The resulting NumPy Array is returned (order of rows unchanged).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pos</strong> (<em>numpy array</em>) – positions to be scaled. Each row is a position.</p></li>
<li><p><strong>scale</strong> (<em>number</em><em> (</em><em>default: 1</em><em>)</em>) – The size of the resulting extent in all directions.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>pos</strong> – scaled positions. Each row is a position.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy array</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rescale_layout_dict</span></code></p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.rotate_xticklabels">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">rotate_xticklabels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">angle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">45</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.rotate_xticklabels" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.rotate_yticklabels">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">rotate_yticklabels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">angle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">45</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.rotate_yticklabels" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.sorted_scatter">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">sorted_scatter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">v</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'RdYlBu_r'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vmax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xlabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ylabel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize_cbar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize_axlabels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize_title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">9</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lim_margin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.sorted_scatter" title="Permalink to this definition"></a></dt>
<dd><p>scatter plot colored with continuous values, with points of the higher
values on the top.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>x</strong> – the x coordinates</p></li>
<li><p><strong>y</strong> – the y coordinates</p></li>
<li><p><strong>v</strong> – the values for each point</p></li>
<li><p><strong>s</strong> – the size of the scatters</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>ax</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.triple_umaps">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">triple_umaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">titles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['dataset',</span> <span class="pre">'is_private',</span> <span class="pre">'is_private']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(6,</span> <span class="pre">4.5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.triple_umaps" title="Permalink to this definition"></a></dt>
<dd><p>Specific function for gene embeddings only, to be generized</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.umap_grid">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">umap_grid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adatas</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ncols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sharex</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sharey</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.umap_grid" title="Permalink to this definition"></a></dt>
<dd><p>multiple-plots of embeddings</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.umap_with_annotates">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">umap_with_annotates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_ids</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anno_df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">text_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">anno_fontsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">plkwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.umap_with_annotates" title="Permalink to this definition"></a></dt>
<dd><p>Plot the UMAP embeddings and annotate the names of the given points</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adt</strong> – </p></li>
<li><p><strong>text_ids</strong> – ids of the points to be annotated, corresponding to
<code class="docutils literal notranslate"><span class="pre">anno_df.index</span></code> or <code class="docutils literal notranslate"><span class="pre">anno_df[index_col]</span></code> (if <code class="docutils literal notranslate"><span class="pre">index_col</span></code> is given)</p></li>
<li><p><strong>color</strong> – a column name in <code class="docutils literal notranslate"><span class="pre">adt.obs</span></code></p></li>
<li><p><strong>anno_df</strong> – use <code class="docutils literal notranslate"><span class="pre">adt.obs</span></code> as the default annotations.</p></li>
<li><p><strong>text_col</strong> – used as the annotation texts.
if specifies, should be a column name in <code class="docutils literal notranslate"><span class="pre">anno_df</span></code>.</p></li>
<li><p><strong>index_col</strong> – used for index the points, corresponding to <code class="docutils literal notranslate"><span class="pre">text_ids</span></code>
if specifies, should be a column name in <code class="docutils literal notranslate"><span class="pre">anno_df</span></code>.</p></li>
<li><p><strong>anno_fontsize</strong> – the fontsize of the annotations</p></li>
<li><p><strong>ax</strong> – the plot ax</p></li>
<li><p><strong>fp</strong> – file path for saving the plot.</p></li>
<li><p><strong>plkwds</strong> – other key-word-args for <code class="docutils literal notranslate"><span class="pre">sc.pl.umap</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>ax</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.venn_plot">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">venn_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">set_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regular</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tt</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Venn</span> <span class="pre">plot'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.venn_plot" title="Permalink to this definition"></a></dt>
<dd><p>sets: iterable
set_labels: list[str]
regular: bool, only for sets of strings!</p>
<blockquote>
<div><p>whether to regularize the strings in sets</p>
</div></blockquote>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.view_color_map">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">view_color_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cmap</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'viridis'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(6,</span> <span class="pre">2)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">grid</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.view_color_map" title="Permalink to this definition"></a></dt>
<dd><p>n: total number of colors
k: number of colors to be plotted on each line.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Set1&#39;</span><span class="p">,</span> <span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="s1">&#39;Spectral&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">view_color_map</span><span class="p">(</span><span class="n">colors</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">n</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cmap</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">palettes</span><span class="o">.</span><span class="n">zeileis_26</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cmap</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">palettes</span><span class="o">.</span><span class="n">default_64</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">view_color_map</span><span class="p">(</span><span class="n">cmap</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">16</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.plot.wrapper_heatmap_scores">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.plot.</span></span><span class="sig-name descname"><span class="pre">wrapper_heatmap_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'celltype'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_pred</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'predicted'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">(5,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_subsample</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Cell</span> <span class="pre">type'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.plot.wrapper_heatmap_scores" title="Permalink to this definition"></a></dt>
<dd><p>sort columns and rows, plot heatmap of cell-type scores</p>
</dd></dl>

</section>
<section id="module-BrainAlign.came.utils.preprocess">
<span id="brainalign-came-utils-preprocess-module"></span><h2>BrainAlign.came.utils.preprocess module<a class="headerlink" href="#module-BrainAlign.came.utils.preprocess" title="Permalink to this heading"></a></h2>
<p>Created on Thu Aug  6 00:21:15 2020
&#64;author: Xingyan Liu</p>
<p>Functions for handling AnnData:</p>
<blockquote>
<div><ul class="simple">
<li><p>I/O functions</p></li>
<li><dl class="simple">
<dt>preprocessing</dt><dd><ul>
<li><p>normalization / z-scoring / …</p></li>
<li><p>groups / …</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>statistics computation</p></li>
</ul>
</div></blockquote>
<dl class="py class">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.AdjacentTrans">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">AdjacentTrans</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vars1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vars2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trans_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.AdjacentTrans" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.AdjacentTrans.T">
<span class="sig-name descname"><span class="pre">T</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.AdjacentTrans.T" title="Permalink to this definition"></a></dt>
<dd><p>Transpose</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.AdjacentTrans.from_edge_df">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_edge_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">edge_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_weight</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.AdjacentTrans.from_edge_df" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.AdjacentTrans.reduce_to_align">
<span class="sig-name descname"><span class="pre">reduce_to_align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'__'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.AdjacentTrans.reduce_to_align" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.AdjacentTrans.reduce_to_align_features">
<span class="sig-name descname"><span class="pre">reduce_to_align_features</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">feats</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.AdjacentTrans.reduce_to_align_features" title="Permalink to this definition"></a></dt>
<dd><p>feats: a sample-by-vars matrix or (sparse) array</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.AdjacentTrans.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.AdjacentTrans.shape" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.adata_from_raw">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">adata_from_raw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dirname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">backup_npz</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_mtx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'matrix'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'float32'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kw</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.adata_from_raw" title="Permalink to this definition"></a></dt>
<dd><p>read matrix and annotated names from directory <cite>dirname</cite></p>
<ul class="simple">
<li><dl class="simple">
<dt>dirname</dt><dd><ul>
<li><p>matrix.npz</p></li>
<li><p>genes.tsv</p></li>
<li><p>barcodes.tsv (or meta_cell.tsv)</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
<p>Alternative (for load mtx file):
&gt;&gt;&gt; adata = sc.read_10x_mtx(dirname)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.add_columns">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">add_columns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_diff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwannos</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.add_columns" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.add_new_edgetype_to_dglgraph">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">add_new_edgetype_to_dglgraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">src_edge_tgt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">tuple</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adj</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">spmatrix</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.add_new_edgetype_to_dglgraph" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.add_obs_annos">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">add_obs_annos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Mapping</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_diff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.add_obs_annos" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.add_var_annos">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">add_var_annos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Mapping</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_diff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.add_var_annos" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.agg_group_edges">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">agg_group_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">asdf</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.agg_group_edges" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adj</strong> – adjacent matrix of shape (N0, N1), if <cite>labels2</cite> is None, then set N0=N1.</p></li>
<li><p><strong>labels1</strong> – a list or a np.array of shape (N0,)</p></li>
<li><p><strong>labels2</strong> – a list or a np.array of shape (N1,)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>summation of connected edges between given groups</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>group_conn</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.align_adata_vars">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">align_adata_vars</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata2</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">df_varmap_1v1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unify_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">[&lt;class</span> <span class="pre">'anndata._core.anndata.AnnData'&gt;,</span> <span class="pre">&lt;class</span> <span class="pre">'anndata._core.anndata.AnnData'&gt;]</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.align_adata_vars" title="Permalink to this definition"></a></dt>
<dd><p>Align the vaiables of two <code class="docutils literal notranslate"><span class="pre">sc.AnnData</span></code> objects</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata1</strong> (<em>AnnData</em>) – reference data</p></li>
<li><p><strong>adata2</strong> (<em>AnnData</em>) – query data</p></li>
<li><p><strong>df_varmap_1v1</strong> – a 2-column DataFrame containing one-to-one mapping between
variables in <code class="docutils literal notranslate"><span class="pre">adata1</span></code> and <code class="docutils literal notranslate"><span class="pre">adata2</span></code></p></li>
<li><p><strong>unify_names</strong> – whether to change the variable-names in <code class="docutils literal notranslate"><span class="pre">adata2</span></code> to match the
<code class="docutils literal notranslate"><span class="pre">adata1.var_names</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.all_vars_of_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">all_vars_of_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.all_vars_of_adata" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.augment_repeat_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">augment_repeat_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_y</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_tot_each</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">concat</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1234</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_prefix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'aug'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.augment_repeat_adata" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.bisplit_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">bisplit_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left_groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reset_index_by</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">None</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.bisplit_adata" title="Permalink to this definition"></a></dt>
<dd><p>Split an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object into a pair of AnnData objects</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>sc.AnnData</em>) – </p></li>
<li><p><strong>key</strong> – a column-name of <cite>adata.obs</cite></p></li>
<li><p><strong>left_groups</strong> – a list of names specifying which groups to be returned as the left one.</p></li>
<li><p><strong>reset_index_by</strong> – the column(s) that store the observation names.
if provided, the observation names will be reset.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>a list of two <code class="docutils literal notranslate"><span class="pre">AnnData</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.change_names">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">change_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">foo_change</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Callable</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.change_names" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>names</strong> – a list of names to be modified</p></li>
<li><p><strong>foo_change</strong> (<em>function to map a name-string to a new one</em>) – </p></li>
<li><p><strong>**kwargs</strong> (<em>other kwargs for foo_change</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.compute_and_get_DEGs">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">compute_and_get_DEGs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unique</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_redo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rank_genes_groups'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">do_normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'t-test'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_info</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cuts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.compute_and_get_DEGs" title="Permalink to this definition"></a></dt>
<dd><p>Compute and get DEGs from <code class="docutils literal notranslate"><span class="pre">sc.AnnData</span></code> object</p>
<p>By default, assume that the counts in adata has been normalized.
If <cite>force_redo</cite>: re-compute DEGs and the original DEGs in adata will be ignored</p>
<p>cuts: dict with keys ‘cut_padj’, ‘cut_pts’, and ‘cut_logfc’</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.compute_unkn_rate">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">compute_unkn_rate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gcnt</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.compute_unkn_rate" title="Permalink to this definition"></a></dt>
<dd><p>unknown rate computing</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.compute_unkn_rate_for2">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">compute_unkn_rate_for2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gcnt1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gcnt2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.compute_unkn_rate_for2" title="Permalink to this definition"></a></dt>
<dd><p>unknown rate computing</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.describe_mat">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">describe_mat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.describe_mat" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.describe_mat_nnz">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">describe_mat_nnz</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.describe_mat_nnz" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.dict_has_keys">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">dict_has_keys</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Mapping</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">keys</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.dict_has_keys" title="Permalink to this definition"></a></dt>
<dd><p>return True if all the keys are in the dict, else False.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.filter_mitogenes">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">filter_mitogenes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.filter_mitogenes" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.get_homologies">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">get_homologies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_match</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vals</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uniquelist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_null</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.get_homologies" title="Permalink to this definition"></a></dt>
<dd><p>Get the homologous gene of input ones based on the homology-mappings</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_match</strong> – homologous gene mappings</p></li>
<li><p><strong>vals</strong> – the gene names.</p></li>
<li><p><strong>cols</strong> – Two column names as <code class="docutils literal notranslate"><span class="pre">[col_from,</span> <span class="pre">col_to]</span></code>, specifying the homology
mapping direction.</p></li>
<li><p><strong>reverse</strong> – whether to reverse the mappings</p></li>
<li><p><strong>uniquelist</strong> – whether to remove the duplicates from the results.</p></li>
<li><p><strong>with_null</strong> – whether to return those genes without any homologies (copy names)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>If <code class="docutils literal notranslate"><span class="pre">uniquelist=True</span></code> and <code class="docutils literal notranslate"><span class="pre">with_null=False</span></code>, a single list will be returned;</p></li>
<li><p><em>otherwise, returns a tuple of two lists (homo, null)</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.get_homologues">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">get_homologues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_match</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vals</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">uniquelist</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_null</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.get_homologues" title="Permalink to this definition"></a></dt>
<dd><p>Get the homologous gene of input ones based on the homology-mappings</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_match</strong> – homologous gene mappings</p></li>
<li><p><strong>vals</strong> – the gene names.</p></li>
<li><p><strong>cols</strong> – Two column names as <code class="docutils literal notranslate"><span class="pre">[col_from,</span> <span class="pre">col_to]</span></code>, specifying the homology
mapping direction.</p></li>
<li><p><strong>reverse</strong> – whether to reverse the mappings</p></li>
<li><p><strong>uniquelist</strong> – whether to remove the duplicates from the results.</p></li>
<li><p><strong>with_null</strong> – whether to return those genes without any homologies (copy names)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>If <code class="docutils literal notranslate"><span class="pre">uniquelist=True</span></code> and <code class="docutils literal notranslate"><span class="pre">with_null=False</span></code>, a single list will be returned;</p></li>
<li><p><em>otherwise, returns a tuple of two lists (homo, null)</em></p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.get_hvgs">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">get_hvgs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_redo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_hvg</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'highly_variable'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">hvg_kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.get_hvgs" title="Permalink to this definition"></a></dt>
<dd><p>packed function for computing and get HVGs from <code class="docutils literal notranslate"><span class="pre">sc.AnnData</span></code> object.
if <cite>force_redo</cite> is True, data should be normalized and log-transformed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – The annotated data matrix of shape <cite>n_obs</cite> × <cite>n_vars</cite>. Rows correspond
to cells and columns to genes.</p></li>
<li><p><strong>force_redo</strong> (<em>bool</em>) – whether to recompute HVGs even when it has already been done</p></li>
<li><p><strong>batch_key</strong> – A column name in <cite>adata.obs</cite></p></li>
<li><p><strong>n_top_genes</strong> – Number of highly-variable genes to keep.
If specified, other parameters will be ignored.</p></li>
<li><p><strong>key_hvg</strong> – A column that should be in <code class="docutils literal notranslate"><span class="pre">adata.var.columns</span></code></p></li>
<li><p><strong>hvg_kwds</strong> – Other parameters, e.g. min_mean, max_mean, min_disp, max_disp</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.get_marker_info_table">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">get_marker_info_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rank_genes_groups'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_padj</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_logfc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_pts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.get_marker_info_table" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.get_marker_name_table">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">get_marker_name_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rank_genes_groups'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.get_marker_name_table" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.get_scnet">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">get_scnet</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.get_scnet" title="Permalink to this definition"></a></dt>
<dd><p>Extract the pre-computed single-cell KNN network</p>
<p>If the adata has not been preprocessed, please run
<cite>adata_processed = quick_preprocess(adata, **kwds)</cite> first.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>adata</strong> – the data object</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.group_mean">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">group_mean</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">spmatrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binary</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.group_mean" title="Permalink to this definition"></a></dt>
<dd><p>compute the group averaged features</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>np.ndarray</em><em> or </em><em>sparse.spmatrix</em>) – shape (n_samples, n_features)</p></li>
<li><p><strong>labels</strong> – shape (n_samples, )</p></li>
<li><p><strong>binary</strong> – if True, the results will be the non-zero proportions</p></li>
<li><p><strong>classes</strong> – optional, names of groups</p></li>
<li><p><strong>features</strong> – optional, names of features</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>average_mat</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.group_mean_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">group_mean_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binary</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.group_mean_adata" title="Permalink to this definition"></a></dt>
<dd><p>Compute averaged feature-values for each group</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – </p></li>
<li><p><strong>groupby</strong> (<em>str</em>) – a column name in adata.obs</p></li>
<li><p><strong>features</strong> – a subset of names in adata.var_names (or adata.raw.var_names)</p></li>
<li><p><strong>binary</strong> (<em>bool</em>) – if True, the results will turn to be the non-zeor proportions for
all (or the given) features</p></li>
<li><p><strong>use_raw</strong> (<em>bool</em>) – whether to access adata.raw to compute the averages.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>a pd.DataFrame with features as index and groups as columns</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.group_mean_dense">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">group_mean_dense</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.group_mean_dense" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.group_mean_multiadata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">group_mean_multiadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adatas</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keys</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tags</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.group_mean_multiadata" title="Permalink to this definition"></a></dt>
<dd><p>tags: tag of each adata to avoid column-name collision</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.group_mean_sparse">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">group_mean_sparse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">spmatrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">print_groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.group_mean_sparse" title="Permalink to this definition"></a></dt>
<dd><p>This function may work with more efficiency than <cite>df.groupby().mean()</cite>
when handling sparse matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>np.ndarray</em><em> or </em><em>sparse.spmatrix</em>) – shape (n_samples, n_features)</p></li>
<li><p><strong>labels</strong> – shape (n_samples, )</p></li>
<li><p><strong>classes</strong> – optional, names of groups</p></li>
<li><p><strong>features</strong> – optional, names of features</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.group_median_dense">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">group_median_dense</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.group_median_dense" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.group_value_counts">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">group_value_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_on</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_by</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">split</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.group_value_counts" title="Permalink to this definition"></a></dt>
<dd><p>extended function of pd.value_counts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>count_on</strong> (<em>str</em>) – which column to apply function <cite>pd.value_counts</cite></p></li>
<li><p><strong>group_by</strong> (<em>str</em>) – which column to group-by</p></li>
<li><p><strong>split</strong> (<em>bool</em>) – split the Series and concatenate them as columns if True.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.group_zscore">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">group_zscore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">ndarray</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_mean</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_value</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.group_zscore" title="Permalink to this definition"></a></dt>
<dd><p>For each column of X, do within-group centering (z-scoring)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>np.ndarray</em><em> or </em><em>pd.DataFrame</em>) – A matrix of shape (n_samples, n_features), each row of X is an
observation, wile each column is a feature</p></li>
<li><p><strong>labels</strong> (<em>np.ndarray</em>) – the group labels, of shape (n_samples,)</p></li>
<li><p><strong>with_mean</strong> (<em>boolean</em><em>, </em><em>True by default</em>) – If True, center the data before scaling, and X shoud be a dense matrix.</p></li>
<li><p><strong>scale</strong> (<em>bool</em>) – whether to scale with standard deviation</p></li>
<li><p><strong>max_value</strong> (<em>float</em>) – if given, the absolute values of the result matrix will be
clipped at this value.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>the scaled data matrix</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.group_zscore_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">group_zscore_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adt</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'batch'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'counts'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_new</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_value</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_mean</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cover</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.group_zscore_adata" title="Permalink to this definition"></a></dt>
<dd><p>Calculate z-scores for each group of observations in an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adt</strong> (<em>AnnData</em>) – </p></li>
<li><p><strong>groupby</strong> (<em>str</em>) – A key from adt.obs, from which the labels are take</p></li>
<li><p><strong>key</strong> (<em>str</em><em>, </em><em>{'X_pca'</em><em>, </em><em>'count'}</em>) – can be a key from adt.obsm, e.g. <cite>key=’X_pca’</cite>
If key == ‘counts’, then do scaling on <cite>adt.X</cite>
and cover the old count matrix, ignoring the <cite>cover</cite> parameter</p></li>
<li><p><strong>key_new</strong> (<em>str</em>) – used when <code class="docutils literal notranslate"><span class="pre">key</span> <span class="pre">!=</span> <span class="pre">'counts'</span></code></p></li>
<li><p><strong>with_mean</strong> (<em>boolean</em><em>, </em><em>True by default</em>) – If True, center the data before scaling, and X shoud be a dense matrix.</p></li>
<li><p><strong>max_value</strong> (<em>float</em>) – if given, the absolute values of the result matrix will be
clipped at this value.</p></li>
<li><p><strong>cover</strong> (<em>bool</em>) – whether to cover the old X with the scored X</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.homograph_from_scipy">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">homograph_from_scipy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adj</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_dgl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">self_loop</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.homograph_from_scipy" title="Permalink to this definition"></a></dt>
<dd><p>making a homologous DGL graph from a scipy.sparse matrix</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.is1v1pairs">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">is1v1pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.is1v1pairs" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.is_symetric">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">is_symetric</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mat</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.is_symetric" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.label_binarize_each">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">label_binarize_each</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">classes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sparse_out</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.label_binarize_each" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.load_dense">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">load_dense</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fpath</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.load_dense" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.load_namelist">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">load_namelist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fpath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">icol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kw</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.load_namelist" title="Permalink to this definition"></a></dt>
<dd><p>icol: int
only return the first column by default</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.load_sparse">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">load_sparse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fpath</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">backup_npz</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.load_sparse" title="Permalink to this definition"></a></dt>
<dd><p><cite>fpath</cite> should be ended with ‘.mtx’ or ‘npz’</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.make_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">make_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mat</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">spmatrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Mapping</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">var</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Mapping</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fn</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assparse</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.make_adata" title="Permalink to this definition"></a></dt>
<dd><p>An alternative way to construct AnnData (BUG fixed).
Something might go wrong when saving the AnnData object constructed
by the build-in function <cite>sc.AnnData</cite>.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.make_bipartite_adj">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">make_bipartite_adj</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_map</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodes1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nodes2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_singleton</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">symmetric</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.make_bipartite_adj" title="Permalink to this definition"></a></dt>
<dd><p>Make a bipartite adjacent (sparse) matrix from a <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code> with
two mapping columns.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_map</strong> (<em>pd.DataFrame with 2 columns.</em>) – each row represent an edge between 2 nodes from the left and right
group of nodes of the bipartite-graph.</p></li>
<li><p><strong>nodes1</strong> (<em>list-like</em>) – node name-list representing left and right nodes, respectively.</p></li>
<li><p><strong>nodes2</strong> (<em>list-like</em>) – node name-list representing left and right nodes, respectively.</p></li>
<li><p><strong>key_data</strong> (<em>str</em><em> or </em><em>None</em>) – if provided, should be a name in df.columns, where the values will be
taken as the weights of the adjacent matrix.</p></li>
<li><p><strong>with_singleton</strong> – whether allow nodes without any neighbors. (default: True)
if nodes1 and nodes2 are not provided, this parameter makes no difference.</p></li>
<li><p><strong>symmetric</strong> – whether make it symmetric, i.e. X += X.T</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bi_adj</span><span class="p">,</span> <span class="n">nodes1</span><span class="p">,</span> <span class="n">nodes2</span> <span class="o">=</span> <span class="n">make_bipartite_adj</span><span class="p">(</span><span class="n">df_map</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.make_dict_from_str">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">make_dict_from_str</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n2id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.make_dict_from_str" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.make_id_name_maps">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">make_id_name_maps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.make_id_name_maps" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>concatenate 2 list of names, and return 2 kinds of maps:</dt><dd><ol class="arabic simple">
<li><p>integer id –&gt; name</p></li>
<li><p>name –&gt; integer id (a tuple of 2 dicts)</p></li>
</ol>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>using 2 dict to map a name to its id for that
– the 2 name-lists may share the same name!
(while the indices are unique)</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.mean_of_nozeros">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">mean_of_nozeros</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.mean_of_nozeros" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mat</strong> (<em>np.arrary</em><em> or </em><em>sparse matrix</em>) – </p></li>
<li><p><strong>axis</strong> (<em>int</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.merge_adata_groups">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">merge_adata_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_lists</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.merge_adata_groups" title="Permalink to this definition"></a></dt>
<dd><p>Merge the given groups into one single group
which is named as <code class="docutils literal notranslate"><span class="pre">'_'.join(groups[i])</span></code> by default.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – </p></li>
<li><p><strong>key</strong> – a column name in <cite>adata.obs</cite></p></li>
<li><p><strong>group_lists</strong> – a list of lists of group names</p></li>
<li><p><strong>new_key</strong> – a column name to be added to <cite>adata.obs</cite></p></li>
<li><p><strong>rename</strong> (<em>bool</em>) – wheter to re-label the groups after merge</p></li>
<li><p><strong>copy</strong> (<em>bool</em>) – </p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">adata_new</span> <span class="o">=</span> <span class="n">merge_adata_groups</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="n">group_lists</span><span class="o">=</span><span class="p">[[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;B&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;E&#39;</span><span class="p">,</span> <span class="s1">&#39;F&#39;</span><span class="p">]],</span> <span class="n">copy</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.merge_adatas">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">merge_adatas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adatas</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Mapping</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">union</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_keys</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dsnames</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_dsname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.merge_adatas" title="Permalink to this definition"></a></dt>
<dd><p>a new adata will be created, use the raw matrix</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adatas</strong> – a list or a dict of AnnData objects.
use data from adata.raw if not None.</p></li>
<li><p><strong>union</strong> (<em>bool</em>) – whether to take union of the features (e.g. genes)</p></li>
<li><p><strong>obs_keys</strong> – column-names from sample annotations to be merged</p></li>
<li><p><strong>dsnames</strong> – names for these datasets, if None, use numbers ranging from
0 to (n_datasets-1)</p></li>
<li><p><strong>key_dsname</strong> – if provided as a str, a column will be added to the merged <cite>obs</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>adata_merged</strong></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">AnnData</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.merge_group_labels">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">merge_group_labels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_lists</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.merge_group_labels" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>labels</strong> – list-like</p></li>
<li><p><strong>group_lists</strong> – a list of lists of group names to be merged</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">merge_group_labels</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">],</span> <span class="p">[[</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="s1">&#39;b&#39;</span><span class="p">],</span> <span class="p">[</span><span class="s1">&#39;c&#39;</span><span class="p">,</span> <span class="s1">&#39;d&#39;</span><span class="p">]])</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.merge_metas">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">merge_metas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adatas</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">obs_keys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.merge_metas" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.merge_named_matrices">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">merge_named_matrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mat_list</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">ndarray</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">csr_matrix</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_lists</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">Sequence</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dsnames</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">union</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.merge_named_matrices" title="Permalink to this definition"></a></dt>
<dd><p>merge matrices and align their columns
This function code is copied from <cite>scanorama.merge_datasets</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mat_list</strong> – a list of cell-by-gene matrices (np.ndarray or sparse.csr_matrix)</p></li>
<li><p><strong>gene_lists</strong> – a list of gene-list corresponding to the columns of each matrix in
<cite>mat_list</cite></p></li>
<li><p><strong>dsnames</strong> – names for these datasets, if None, use numbers ranging from
0 to (n_datasets-1)</p></li>
<li><p><strong>union</strong> (<em>bool</em>) – whether to take union of the features from all the matrices</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>mat_list</em> – a list of cell-by-gene matrices (sparse.csr_matrix) with aligned gene
columns.</p></li>
<li><p><em>ret_genes</em> – a gene list shared by all the mat_list</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.mtx2df">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">mtx2df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.mtx2df" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.normalize_col">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">normalize_col</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sum'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.normalize_col" title="Permalink to this definition"></a></dt>
<dd><p>make the column elements of X to unit sum</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> – a (sparse) matrix</p></li>
<li><p><strong>scale_factor</strong> (<em>float</em><em> or </em><em>None</em>) – if None, use the median of sum level as the scaling factor.</p></li>
<li><p><strong>by</strong> (<em>str</em><em>, </em><em>{'sum'</em><em>, </em><em>'max'}</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.normalize_default">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">normalize_default</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_return</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.normalize_default" title="Permalink to this definition"></a></dt>
<dd><p>Normalizing datasets with default settings (total-counts normalization
followed by log(x+1) transform).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object</p></li>
<li><p><strong>target_sum</strong> – scale factor of total-count normalization</p></li>
<li><p><strong>copy</strong> – whether to copy the dataset</p></li>
<li><p><strong>log_only</strong> – whether to skip the “total-counts normalization” and only perform
log(x+1) transform</p></li>
<li><p><strong>force_return</strong> – whether to return the data, even if changes are made for the
original object</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">AnnData</span></code> or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.normalize_log_then_total">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">normalize_log_then_total</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_return</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.normalize_log_then_total" title="Permalink to this definition"></a></dt>
<dd><p>For SplitSeq data, performing log(x+1) BEFORE total-sum normalization
will results a better UMAP visualization (e.g. clusters would be less
confounded by different total-counts ).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.normalize_max">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">normalize_max</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.normalize_max" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.normalize_maxmin">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">normalize_maxmin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-08</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.normalize_maxmin" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.normalize_norms">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">normalize_norms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sum'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.normalize_norms" title="Permalink to this definition"></a></dt>
<dd><p>wrapper of <cite>normalize_colsum</cite> and <cite>normalize_rowsum</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> – a (sparse) matrix</p></li>
<li><p><strong>scale_factor</strong> (<em>numeric</em><em>, </em><em>None</em>) – if None, use the median of sum level as the scaling factor.</p></li>
<li><p><strong>axis</strong> (<em>int</em><em>, </em><em>{0</em><em>, </em><em>1}</em>) – if axis = 0, apply to each column;
if axis = 1, apply to each row.</p></li>
<li><p><strong>by</strong> (<em>str</em><em>, </em><em>{'sum'</em><em>, </em><em>'max'}</em>) – normalization method</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.normalize_row">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">normalize_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_factor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sum'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.normalize_row" title="Permalink to this definition"></a></dt>
<dd><p>make the row elements of X to unit sum</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> – a (sparse) matrix</p></li>
<li><p><strong>scale_factor</strong> (<em>float</em><em> or </em><em>None</em>) – if None, use the median of sum level as the scaling factor.</p></li>
<li><p><strong>by</strong> (<em>str</em><em>, </em><em>{'sum'</em><em>, </em><em>'max'}</em>) – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.order_contingency_mat">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">order_contingency_mat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mat</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.order_contingency_mat" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.pivot_df_to_sparse">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">pivot_df_to_sparse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.pivot_df_to_sparse" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>row, col:</dt><dd><p>str or int, int for column index, and str for column name</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.pivot_to_sparse">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">pivot_to_sparse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">rows</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rownames</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colnames</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.pivot_to_sparse" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rownames</strong> – If provided, the resulting matrix rows are restricted to <cite>rownames</cite>,
Names will be removed if they are not in <cite>rownames</cite>, and names that
not occur in <cite>rows</cite> but in <cite>rownames</cite> will take ALL-zeros in that row
of the resulting matrix.
if not provided, will be set as <cite>rows.unique()</cite></p></li>
<li><p><strong>colnames</strong> – if not provided, will be set as <cite>cols.unique()</cite></p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p><cite>rows</cite> and <cite>cols</cite> should be of the same length!</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.quick_pre_clust">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">quick_pre_clust</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hvgs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pcs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nneigh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reso</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">hvg_kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.quick_pre_clust" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.quick_pre_vis">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">quick_pre_vis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hvgs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pcs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nneigh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'umap'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cosine'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">hvg_kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.quick_pre_vis" title="Permalink to this definition"></a></dt>
<dd><p>Go through the default pipeline and have a overall visualization of
the data.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.quick_preprocess">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">quick_preprocess</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hvgs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalize_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_sum</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pcs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nneigh</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">metric</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cosine'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">hvg_kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.quick_preprocess" title="Permalink to this definition"></a></dt>
<dd><p>Quick preprocess of the raw data.</p>
<p class="rubric">Notes</p>
<p>if <cite>normalize_data</cite> is True, an adata with RAW counts is required!</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.reduce_to_align">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">reduce_to_align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adj</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">spmatrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'__'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.reduce_to_align" title="Permalink to this definition"></a></dt>
<dd><p>reduce column-names to align rows</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.regu_gname">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">regu_gname</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pattern</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'^X[0-9]{2,3}'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.regu_gname" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.remove_adata_groups">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">remove_adata_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.remove_adata_groups" title="Permalink to this definition"></a></dt>
<dd><p>Remove given groups from an AnnData object</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.remove_adata_small_groups">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">remove_adata_small_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.remove_adata_small_groups" title="Permalink to this definition"></a></dt>
<dd><p>return adata with small groups removed, grouped by <cite>key</cite></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.remove_small_groups">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">remove_small_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_samples</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indicate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.remove_small_groups" title="Permalink to this definition"></a></dt>
<dd><p>return labels with small groups removed</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.reverse_dict">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">reverse_dict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Mapping</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Mapping</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.reverse_dict" title="Permalink to this definition"></a></dt>
<dd><p>the values of the dict must be un-changable type</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.reverse_dict_of_list">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">reverse_dict_of_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Mapping</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Mapping</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.reverse_dict_of_list" title="Permalink to this definition"></a></dt>
<dd><p>the values of the dict must be list-like type</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.save_mtx2df">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">save_mtx2df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.save_mtx2df" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.save_named_mtx">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">save_named_mtx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dirname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">backup_npz</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.save_named_mtx" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.save_namelist">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">save_namelist</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kw</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.save_namelist" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.scipy_edge_dict_for_dgl">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">scipy_edge_dict_for_dgl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">edge_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">foo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.scipy_edge_dict_for_dgl" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>foo: a callable object.</dt><dd><p>For example, <cite>torch.LongTensor</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.split_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">split_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">AnnData</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.split_adata" title="Permalink to this definition"></a></dt>
<dd><p>Split an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object into a dict of multiple objects</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – the AnnData object to be split</p></li>
<li><p><strong>key</strong> – a column-name of <cite>adata.obs</cite></p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>a dict of AnnData</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.subset_matches">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">subset_matches</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_match</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">right</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">union</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indicators</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.subset_matches" title="Permalink to this definition"></a></dt>
<dd><p>Take a subset of token matches (e.g., gene homologies)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_match</strong> (<em>pd.DataFrame</em>) – a dataframe with at least 2 columns</p></li>
<li><p><strong>left</strong> – list-like, for extracting elements in the first column.</p></li>
<li><p><strong>right</strong> – list-like, for extracting elements in the second column.</p></li>
<li><p><strong>union</strong> – whether to take union of both sides of genes</p></li>
<li><p><strong>cols</strong> – if specified, should be 2 column names for left and right genes,
respectively.</p></li>
<li><p><strong>indicators</strong> – if True, only return the indicators.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.take_1v1_matches">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">take_1v1_matches</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.take_1v1_matches" title="Permalink to this definition"></a></dt>
<dd><p>Take the one-to-one matches of the given two columns of a <code class="docutils literal notranslate"><span class="pre">pd.DataFrame</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – a dataframe with at least two columns</p></li>
<li><p><strong>cols</strong> – column names of <code class="docutils literal notranslate"><span class="pre">df</span></code>, specifying the columns to match.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>A pd.DataFrame contains only one-to-one mappings and the corresponding rows</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.take_adata_groups">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">take_adata_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">onlyx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.take_adata_groups" title="Permalink to this definition"></a></dt>
<dd><p>Take given groups from an AnnData object</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.take_freq1">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">take_freq1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.take_freq1" title="Permalink to this definition"></a></dt>
<dd><p>take elements that occur only once in <cite>lst</cite></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.take_freq1more">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">take_freq1more</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.take_freq1more" title="Permalink to this definition"></a></dt>
<dd><p>take elements that occur more than once in <cite>lst</cite></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.take_group_labels">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">take_group_labels</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">group_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">indicate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.take_group_labels" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>labels</strong> (<em>list-like</em>) – </p></li>
<li><p><strong>group_names</strong> – names of groups that you want to take out</p></li>
<li><p><strong>indicate</strong> (<em>bool</em>) – if True, return a Series of bool indicators of the groups
else, return the labels.</p></li>
<li><p><strong>remove</strong> (<em>bool</em>) – False by default, set as True if you want to keep groups that
NOT in the <cite>given group_names</cite></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.top_markers_from_adata">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">top_markers_from_adata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unique</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_padj</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_logfc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cut_pts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'rank_genes_groups'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.top_markers_from_adata" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.top_markers_from_df">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">top_markers_from_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">marker_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unique</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.top_markers_from_df" title="Permalink to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>marker_df</strong> – a data-frame with cluster names as columns, and genes as values</p></li>
<li><p><strong>n</strong> (<em>int</em>) – </p></li>
<li><p><strong>groups</strong> – a list of cluster names (column names)</p></li>
<li><p><strong>unique</strong> (<em>bool</em>) – whether to flatten the results into a unique gene-list, default is True.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>A dataframe (<code class="docutils literal notranslate"><span class="pre">union=False</span></code>) or a flattened marker list (<code class="docutils literal notranslate"><span class="pre">union=True</span></code>)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.top_markers_from_info">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">top_markers_from_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_info</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unique</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_group</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'group'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'names'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">list</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.top_markers_from_info" title="Permalink to this definition"></a></dt>
<dd><p>df_info: DEG-info table that can be take from <cite>top_markers_from_adata</cite></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.unpack_dict_of_lists">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">unpack_dict_of_lists</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">d</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.unpack_dict_of_lists" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.wrapper_normalize">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">wrapper_normalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'maxmin'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.wrapper_normalize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.wrapper_scale">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">wrapper_scale</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.wrapper_scale" title="Permalink to this definition"></a></dt>
<dd><p>Wrapper function for centering and scaling data matrix <cite>X</cite> in sc.AnnData,
extended for within-batch processing.</p>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">wrapper_scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;batch&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.preprocess.zscore">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.preprocess.</span></span><span class="sig-name descname"><span class="pre">zscore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">with_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.preprocess.zscore" title="Permalink to this definition"></a></dt>
<dd><p>For each column of X, do centering (z-scoring)</p>
</dd></dl>

</section>
<section id="module-BrainAlign.came.utils.train">
<span id="brainalign-came-utils-train-module"></span><h2>BrainAlign.came.utils.train module<a class="headerlink" href="#module-BrainAlign.came.utils.train" title="Permalink to this heading"></a></h2>
<p>Created on Sun Apr 11 19:15:45 2021</p>
<p>&#64;author: Xingyan Liu</p>
<dl class="py class">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.Trainer">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.train.</span></span><span class="sig-name descname"><span class="pre">Trainer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feat_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Mapping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DGLHeteroGraph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_idx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_idx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tensor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Tensor</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">l2norm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dir_main</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">WindowsPath('.')</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.Trainer" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">BaseTrainer</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.Trainer.evaluate_metrics">
<span class="sig-name descname"><span class="pre">evaluate_metrics</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">train_preds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_preds</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">train_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.Trainer.evaluate_metrics" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.Trainer.get_current_outputs">
<span class="sig-name descname"><span class="pre">get_current_outputs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">feat_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">g</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">other_inputs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.Trainer.get_current_outputs" title="Permalink to this definition"></a></dt>
<dd><p>get the current states of the model output</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.Trainer.log_info">
<span class="sig-name descname"><span class="pre">log_info</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">train_acc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_acc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ami</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">print_info</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.Trainer.log_info" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.Trainer.plot_class_accs">
<span class="sig-name descname"><span class="pre">plot_class_accs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.Trainer.plot_class_accs" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.Trainer.plot_class_losses">
<span class="sig-name descname"><span class="pre">plot_class_losses</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.Trainer.plot_class_losses" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.Trainer.plot_cluster_index">
<span class="sig-name descname"><span class="pre">plot_cluster_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.Trainer.plot_cluster_index" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.Trainer.set_class_weights">
<span class="sig-name descname"><span class="pre">set_class_weights</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">class_weights=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">foo=&lt;ufunc</span> <span class="pre">'sqrt'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_add=0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.Trainer.set_class_weights" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.Trainer.train">
<span class="sig-name descname"><span class="pre">train</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_epochs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">350</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_class_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params_lossfunc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pass</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.005</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cat_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">info_stride</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">backup_stride</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">222</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">other_inputs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.Trainer.train" title="Permalink to this definition"></a></dt>
<dd><p>Main function for model training (whole-graph based)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n_epochs</strong> (<em>int</em>) – The total number of epochs to train the model</p></li>
<li><p><strong>use_class_weights</strong> (<em>bool</em>) – whether to use the class-weights, useful for unbalanced
sample numbers for each class</p></li>
<li><p><strong>params_lossfunc</strong> (<em>dict</em>) – parameters for loss-functions</p></li>
<li><p><strong>n_pass</strong> (<em>int</em>) – The number of epochs to be skipped (not backup model checkpoint)</p></li>
<li><p><strong>eps</strong> – tolerance for cluster-index</p></li>
<li><p><strong>cat_class</strong> (<em>str</em>) – node type for classification</p></li>
<li><p><strong>device</strong> – one of {‘cpu’, ‘gpu’, None}</p></li>
<li><p><strong>info_stride</strong> (<em>int</em>) – epoch-strides for printing out the training information</p></li>
<li><p><strong>backup_stride</strong> (<em>int</em>) – saving checkpoint after <cite>backup_stride</cite> epochs</p></li>
<li><p><strong>other_inputs</strong> – other inputs for <cite>model.forward()</cite></p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.Trainer.train_minibatch">
<span class="sig-name descname"><span class="pre">train_minibatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_epochs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_class_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params_lossfunc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_pass</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cat_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">128</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sampler</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">backup_stride</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">111</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">info_stride</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">other_inputs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.Trainer.train_minibatch" title="Permalink to this definition"></a></dt>
<dd><p>Main function for model training (based on mini-batches)</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.infer_for_nodes">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.train.</span></span><span class="sig-name descname"><span class="pre">infer_for_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">feat_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dataloader</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ntype</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'cell'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">device</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reorder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">argmax_dim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_training</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.infer_for_nodes" title="Permalink to this definition"></a></dt>
<dd><p>“Assume that the model output is a dict of Tensors</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.make_class_weights">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.train.</span></span><span class="sig-name descname"><span class="pre">make_class_weights</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">astensor=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">foo=&lt;ufunc</span> <span class="pre">'sqrt'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_add=0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.make_class_weights" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.order_by_ids">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.train.</span></span><span class="sig-name descname"><span class="pre">order_by_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ids</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.order_by_ids" title="Permalink to this definition"></a></dt>
<dd><p>reorder by the original ids</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.prepare4train">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.train.</span></span><span class="sig-name descname"><span class="pre">prepare4train</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dpair</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="BrainAlign.came.datapair.html#BrainAlign.came.datapair.unaligned.DataPair" title="BrainAlign.came.datapair.unaligned.DataPair"><span class="pre">DataPair</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="BrainAlign.came.datapair.html#BrainAlign.came.datapair.aligned.AlignedDataPair" title="BrainAlign.came.datapair.aligned.AlignedDataPair"><span class="pre">AlignedDataPair</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_class</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_clust</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'clust_lbs'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_within</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_keys</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unit_var</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clip_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(-3,</span> <span class="pre">3.5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">categories</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_labels</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_idx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ground_truth</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node_cls_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cell'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'label'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwds</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">dict</span></span></span><a class="headerlink" href="#BrainAlign.came.utils.train.prepare4train" title="Permalink to this definition"></a></dt>
<dd><p>dpair: DataPair
batch_keys:</p>
<blockquote>
<div><p>a list of two strings (or None), specifying the batch-keys for
data1 and data2, respectively.
If given, features (of cell nodes) will be scaled within each batch</p>
<blockquote>
<div><p>e.g., [‘batch’, ‘sample’]</p>
</div></blockquote>
</div></blockquote>
<dl class="simple">
<dt>test_idx:</dt><dd><p>By default, the testing indices will be decided automatically.
if provided, should be an index-sequence of the same length as
<cite>cluster_labels</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="BrainAlign.came.utils.train.seed_everything">
<span class="sig-prename descclassname"><span class="pre">BrainAlign.came.utils.train.</span></span><span class="sig-name descname"><span class="pre">seed_everything</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">123</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#BrainAlign.came.utils.train.seed_everything" title="Permalink to this definition"></a></dt>
<dd><p>not works well</p>
</dd></dl>

</section>
<section id="brainalign-came-utils-train-v0-module">
<h2>BrainAlign.came.utils.train_v0 module<a class="headerlink" href="#brainalign-came-utils-train-v0-module" title="Permalink to this heading"></a></h2>
</section>
<section id="module-BrainAlign.came.utils">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-BrainAlign.came.utils" title="Permalink to this heading"></a></h2>
<p>Created on Sun Dec 27 21:59:44 2020</p>
<p>&#64;author: Xingyan Liu</p>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Biao Zhang.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>